---
title: "Comparative Genomics Southern Flounder"
output:
    html_document:
    toc: yes
---

```{r load libraries, message=FALSE, warning=FALSE}

knitr::opts_chunk$set(warning = FALSE, message = FALSE)

source("lib/libraries.R")
source("lib/ggplot.R")

```

# Compare family maps I

Determine if there are ambivalent marker placings in each family map (e.g. large gaps). Check segregation distortion/missing data/recombination rates. Create list of markers to remove from each map.

```{r}

# load family map (with duplicate markers added back in)
FamA <- read.table("results/FamA.fullmap",
                   stringsAsFactors = FALSE, sep = "", skip = 1,
                   col.names = c("LOCUS", "LG_FamA","POS_FamA"))

FamB <- read.table("results/FamB.fullmap",
                   stringsAsFactors = FALSE, sep = "", skip = 1,
                   col.names = c("LOCUS", "LG_FamB","POS_FamB"))

temp <- inner_join(FamA, FamB, by = "LOCUS")

write.table(temp, "results/shared.markers",
            quote = FALSE, col.names = TRUE, row.names = FALSE)

```

Identify markers that are out of sequence between the two family maps.

Compare locations in family maps. Determine if in cluster of loci that are not properly resolved in just one of the maps, look at segregation distortion/crossovers/missing data.

Create set of markers to be removed from each map.

Remove from family maps and re-map each one.

## Family A

```{r}

# change header names to Marker | Markertype | Genotype
FamA.onemap <- read.delim("data/LINKMAP/FamA.onemap",
                          header = FALSE, sep = "\t",
                          col.names = c("LOCUS", "LOCUSTYPE", "GENOTYPE"),
                          skip = 1, stringsAsFactors = FALSE)

# number of individuals in dataset
N <- read.delim("data/LINKMAP/FamA.onemap", header = FALSE,
                          col.names = c("N_IND", "N_LOCI"),
                          nrows = 1, stringsAsFactors = FALSE)

# load file with mapped markers and filter onemap file
mapped_markers <- read.table("results/FamA_allmarkers.txt", 
                             header = TRUE, stringsAsFactors = TRUE) %>%
    select(LOCUS) %>%
    transmute(LOCUS = paste(c("*"), LOCUS, sep = ""))

FamA.onemap.filtered <- semi_join(FamA.onemap, mapped_markers)

# remove flagged markers
FamA.onemap.filtered <- FamA.onemap.filtered %>%
    filter(!LOCUS %in% c("*dDocent_Contig_30805",
                         "*dDocent_Contig_248473",
                         "*dDocent_Contig_244117",
                         "*dDocent_Contig_11546",
                         "*dDocent_Contig_35991",
                         "*dDocent_Contig_35624",
                         "*dDocent_Contig_200224",
                         "*dDocent_Contig_7091",
                         "*dDocent_Contig_2358",
                         "*dDocent_Contig_23388",
                         "*dDocent_Contig_186256",
                         "*dDocent_Contig_3817",
                         "*dDocent_Contig_97890",
                         "*dDocent_Contig_298306",
                         "*dDocent_Contig_67843",
                         "*dDocent_Contig_29969",
                         "*dDocent_Contig_15670",
                         "*dDocent_Contig_30792",
                         "*dDocent_Contig_75671",
                         "*dDocent_Contig_172925",
                         "*dDocent_Contig_144097",
                         "*dDocent_Contig_64257",
                         "*dDocent_Contig_12459",
                         "*dDocent_Contig_44930",
                         "*dDocent_Contig_990",
                         "*dDocent_Contig_50993",
                         "*dDocent_Contig_8255",
                         "*dDocent_Contig_20685",
                         "*dDocent_Contig_57614",
                         "*dDocent_Contig_3486",
                         "*dDocent_Contig_69118",
                         "*dDocent_Contig_51033",
                         "*dDocent_Contig_34901",
                         "*dDocent_Contig_1452",
                         "*dDocent_Contig_2847",
                         "*dDocent_Contig_6716",
                         "*dDocent_Contig_60349",
                         "*dDocent_Contig_77680",
                         "*dDocent_Contig_12222",
                         "*dDocent_Contig_62798",
                         "*dDocent_Contig_87392",
                         "*dDocent_Contig_39281",
                         "*dDocent_Contig_13012",
                         "*dDocent_Contig_44363",
                         "*dDocent_Contig_52542",
                         "*dDocent_Contig_123370",
                         "*dDocent_Contig_63085",
                         "*dDocent_Contig_138814",
                         "*dDocent_Contig_261231",
                         "*dDocent_Contig_47638",
                         "*dDocent_Contig_9761",
                         "*dDocent_Contig_4017",
                         "*dDocent_Contig_233860",
                         "*dDocent_Contig_24796",
                         "*dDocent_Contig_30828",
                         "*dDocent_Contig_52309",
                         "*dDocent_Contig_78689",
                         "*dDocent_Contig_62261",
                         "*dDocent_Contig_97106",
                         "*dDocent_Contig_1694",
                         "*dDocent_Contig_184316",
                         "*dDocent_Contig_483",
                         "*dDocent_Contig_596",
                         "*dDocent_Contig_13530",
                         "*dDocent_Contig_14522",
                         "*dDocent_Contig_50994",
                         "*dDocent_Contig_315477",
                         "*dDocent_Contig_730",
                         "*dDocent_Contig_107875",
                         "*dDocent_Contig_41408",
                         '*dDocent_Contig_213918',
                         "*dDocent_Contig_130476",
                         "*dDocent_Contig_129129",
                         "*dDocent_Contig_69979",
                         "*dDocent_Contig_224858",
                         "*dDocent_Contig_14784",
                         "*dDocent_Contig_17546",
                         "*dDocent_Contig_169675",
                         "*dDocent_Contig_26617",
                         "*dDocent_Contig_180953",
                         "*dDocent_Contig_16379",
                         "*dDocent_Contig_333713",
                         "*dDocent_Contig_52478",
                         "*dDocent_Contig_294562",
                         "*dDocent_Contig_17020",
                         "*dDocent_Contig_29858",
                         "*dDocent_Contig_6740",
                         "*dDocent_Contig_59297",
                         "*dDocent_Contig_3323",
                         "*dDocent_Contig_104837",
                         "*dDocent_Contig_305075",
                         "*dDocent_Contig_52003",
                         "*dDocent_Contig_29931",
                         "*dDocent_Contig_10018",
                         "*dDocent_Contig_65233",
                         "*dDocent_Contig_2867",
                         "*dDocent_Contig_42394",
                         "*dDocent_Contig_360",
                         "*dDocent_Contig_25067"))

# remove duplicate markers
temp <- FamA.onemap.filtered %>%
    distinct(GENOTYPE, .keep_all = TRUE)
    
# create dataframe of removed loci and corresponding loci
match_locA <- anti_join(FamA.onemap.filtered, temp) %>%
  rename(MATCHLOC = LOCUS) %>%
  select(MATCHLOC, GENOTYPE) %>%
  left_join(temp, by = "GENOTYPE") %>%
  select(LOCUS, MATCHLOC)

# write filtered input file for Onemap
# need to add extra line back in
con <- file("data/LINKMAP/FamAfilter.onemap", open = "wt")
writeLines(paste(N$N_IND, (nrow(FamA.onemap.filtered))), con)
write.table(FamA.onemap.filtered, con,
            append = FALSE, quote = FALSE, sep = " ",
            na = "NA", dec = ".",
            col.names = FALSE, row.names = FALSE)
close(con)

# import datafile
FamA.Map <- read.outcross(file="data/LINKMAP/FamAfilter.onemap")

# Summary of datafile
FamA.Map

# estimate recombination fraction between all pairs of markers
# creates object of class rf.2pts
FamA.Map.rf.L6.rf35 <- rf.2pts(FamA.Map, LOD = 6, max.rf = 0.35, verbose=TRUE)

# create sequence (list) of all markers to assign to LGs
all.markers <- make.seq(FamA.Map.rf.L6.rf35, "all")

# assign markers to linkage groups
LGs <- group(all.markers, LOD = 6, max.rf = 0.35)

# define map function
set.map.fun(typ = "haldane")

# create list of separate LGs
# (convert from class group to class sequence)
LG_list <- lapply(c(1:24), function(LG) {
  make.seq(LGs, LG)
})

# uses compare() to create a framework of most informative markers
# map remaining markers using try.seq()
# this step will take a few hours
LG_ordered <- lapply((LG_list), function(LG){
  order.seq(LG, n.init=6, THRES=3, touchdown=TRUE, tol=10e-2)
})

# create marker order with all markers
LG_final <- lapply((LG_ordered), function(LG){
  make.seq(LG, "force")
})

# group all final ordered LGs
FamA.map <- LG_final

# draw map for all LGs
draw.map(FamA.map, names=FALSE, grid=TRUE, cex.mrk=0.7)

# export map to file
write.map(FamA.map, "data/LINKMAP/FamA.fullmap")

FamA.fullmap <- read.table("data/LINKMAP/FamA.fullmap",
                           skip = 1, header = FALSE) %>%
  mutate(LOCUS = paste(c("*"), V2, sep = "")) %>%
  rename(LG = V1, POSITION = V3) %>%
  select(LOCUS, LG, POSITION)

# add mapping positions
temp <- left_join(match_locA, FamA.fullmap, by = "LOCUS") %>%
  select(MATCHLOC, LG, POSITION) %>%
  rename(LOCUS = MATCHLOC)

# add markers to final map
FamA.fullmap <- bind_rows(FamA.fullmap, temp) %>%
  group_by(LG) %>%
  arrange(POSITION) %>%
  ungroup()

# write output to file
write.table(FamA.fullmap, "results/FamA.fullmap-1", row.names = FALSE)

```

## Family B:

```{r}

# change header names to Marker | Markertype | Genotype
FamB.onemap <- read.delim("data/LINKMAP/FamB.onemap",
                          header = FALSE, sep = "\t",
                          col.names = c("LOCUS", "LOCUSTYPE", "GENOTYPE"),
                          skip = 1, stringsAsFactors = FALSE)

# number of individuals in dataset
N <- read.delim("data/LINKMAP/FamB.onemap", header = FALSE,
                          col.names = c("N_IND", "N_LOCI"),
                          nrows = 1, stringsAsFactors = FALSE)

# load file with mapped markers and filter onemap file
mapped_markers <- read.table("results/FamB_allmarkers.txt", 
                             header = TRUE, stringsAsFactors = TRUE) %>%
    select(LOCUS) %>%
    transmute(LOCUS = paste(c("*"), LOCUS, sep = ""))

FamB.onemap.filtered <- semi_join(FamB.onemap, mapped_markers)

# remove flagged markers
FamB.onemap.filtered <- FamB.onemap.filtered %>%
  filter(!LOCUS %in% c("*dDocent_Contig_30805",
                       "*dDocent_Contig_248473",
                       "*dDocent_Contig_244117",
                       "*dDocent_Contig_11546",
                       "*dDocent_Contig_35991",
                       "*dDocent_Contig_35624",
                       "*dDocent_Contig_200224",
                       "*dDocent_Contig_7091",
                       "*dDocent_Contig_2358",
                       "*dDocent_Contig_23388",
                       "*dDocent_Contig_186256",
                       "*dDocent_Contig_3817",
                       "*dDocent_Contig_97890",
                       "*dDocent_Contig_298306",
                       "*dDocent_Contig_67843",
                       "*dDocent_Contig_29969",
                       "*dDocent_Contig_15670",
                       "*dDocent_Contig_30792",
                       "*dDocent_Contig_75671",
                       "*dDocent_Contig_172925",
                       "*dDocent_Contig_144097",
                       "*dDocent_Contig_64257",
                       "*dDocent_Contig_12459",
                       "*dDocent_Contig_44930",
                       "*dDocent_Contig_990",
                       "*dDocent_Contig_50993",
                       "*dDocent_Contig_8255",
                       "*dDocent_Contig_20685",
                       "*dDocent_Contig_57614",
                       "*dDocent_Contig_3486",
                       "*dDocent_Contig_69118",
                       "*dDocent_Contig_51033",
                       "*dDocent_Contig_34901",
                       "*dDocent_Contig_1452",
                       "*dDocent_Contig_2847",
                       "*dDocent_Contig_6716",
                       "*dDocent_Contig_60349",
                       "*dDocent_Contig_77680",
                       "*dDocent_Contig_12222",
                       "*dDocent_Contig_62798",
                       "*dDocent_Contig_87392",
                       "*dDocent_Contig_39281",
                       "*dDocent_Contig_13012",
                       "*dDocent_Contig_44363",
                       "*dDocent_Contig_52542",
                       "*dDocent_Contig_123370",
                       "*dDocent_Contig_63085",
                       "*dDocent_Contig_138814",
                       "*dDocent_Contig_261231",
                       "*dDocent_Contig_47638",
                       "*dDocent_Contig_9761",
                       "*dDocent_Contig_4017",
                       "*dDocent_Contig_233860",
                       "*dDocent_Contig_24796",
                       "*dDocent_Contig_30828",
                       "*dDocent_Contig_52309",
                       "*dDocent_Contig_78689",
                       "*dDocent_Contig_62261",
                       "*dDocent_Contig_97106",
                       "*dDocent_Contig_1694",
                       "*dDocent_Contig_184316",
                       "*dDocent_Contig_483",
                       "*dDocent_Contig_596",
                       "*dDocent_Contig_13530",
                       "*dDocent_Contig_14522",
                       "*dDocent_Contig_50994",
                       "*dDocent_Contig_315477",
                       "*dDocent_Contig_730",
                       "*dDocent_Contig_107875",
                       "*dDocent_Contig_41408",
                       "*dDocent_Contig_213918",
                       "*dDocent_Contig_130476",
                       "*dDocent_Contig_129129",
                       "*dDocent_Contig_69979",
                       "*dDocent_Contig_224858",
                       "*dDocent_Contig_14784",
                       "*dDocent_Contig_17546",
                       "*dDocent_Contig_169675",
                       "*dDocent_Contig_26617",
                       "*dDocent_Contig_180953",
                       "*dDocent_Contig_16379",
                       "*dDocent_Contig_333713",
                       "*dDocent_Contig_52478",
                       "*dDocent_Contig_294562",
                       "*dDocent_Contig_17020",
                       "*dDocent_Contig_29858",
                       "*dDocent_Contig_6740",
                       "*dDocent_Contig_59297",
                       "*dDocent_Contig_3323",
                       "*dDocent_Contig_104837",
                       "*dDocent_Contig_305075",
                       "*dDocent_Contig_52003",
                       "*dDocent_Contig_29931",
                       "*dDocent_Contig_10018",
                       "*dDocent_Contig_65233",
                       "*dDocent_Contig_2867",
                       "*dDocent_Contig_42394",
                       "*dDocent_Contig_360",
                       "*dDocent_Contig_25067"))

# remove duplicate markers
temp <- FamB.onemap.filtered %>%
    distinct(GENOTYPE, .keep_all = TRUE)
    
# create dataframe of removed loci and corresponding loci
match_locB <- anti_join(FamB.onemap.filtered, temp) %>%
  rename(MATCHLOC = LOCUS) %>%
  select(MATCHLOC, GENOTYPE) %>%
  left_join(temp, by = "GENOTYPE") %>%
  select(LOCUS, MATCHLOC)

# write filtered input file for Onemap
# need to add extra line back in
con <- file("data/LINKMAP/FamBfilter.onemap", open = "wt")
writeLines(paste(N$N_IND, (nrow(FamB.onemap.filtered))), con)
write.table(FamB.onemap.filtered, con,
            append = FALSE, quote = FALSE, sep = " ",
            na = "NA", dec = ".",
            col.names = FALSE, row.names = FALSE)
close(con)

# import datafile
FamB.Map <- read.outcross(file="data/LINKMAP/FamBfilter.onemap")

# Summary of datafile
FamB.Map

# estimate recombination fraction between all pairs of markers
# creates object of class rf.2pts
FamB.Map.rf.L6.rf35 <- rf.2pts(FamB.Map, LOD = 6, max.rf = 0.35, verbose=TRUE)

# create sequence (list) of all markers to assign to LGs
all.markers <- make.seq(FamB.Map.rf.L6.rf35, "all")

# assign markers to linkage groups
LGs <- group(all.markers, LOD = 6, max.rf = 0.35)

# define map function
set.map.fun(typ = "haldane")

# create list of separate LGs
# (convert from class group to class sequence)
LG_list <- lapply(c(1:24), function(LG) {
  make.seq(LGs, LG)
})

# uses compare() to create a framework of most informative markers
# map remaining markers using try.seq()
# this step will take a few hours
LG_ordered <- lapply((LG_list), function(LG){
  order.seq(LG, n.init=6, THRES=3, touchdown=TRUE, tol=10e-2)
})

# create marker order with all markers
LG_final <- lapply((LG_ordered), function(LG){
  make.seq(LG, "force")
})

# group all final ordered LGs
FamB.map <- LG_final

# draw map for all LGs
draw.map(FamB.map, names=FALSE, grid=TRUE, cex.mrk=0.7)

# export map to file
write.map(FamB.map, "data/LINKMAP/FamB.fullmap")

FamB.fullmap <- read.table("data/LINKMAP/FamB.fullmap",
                           skip = 1, header = FALSE) %>%
  mutate(LOCUS = paste(c("*"), V2, sep = "")) %>%
  rename(LG = V1, POSITION = V3) %>%
  select(LOCUS, LG, POSITION)

# add mapping positions
temp <- left_join(match_locB, FamB.fullmap, by = "LOCUS") %>%
  select(MATCHLOC, LG, POSITION) %>%
  rename(LOCUS = MATCHLOC)

# add markers to final map
FamB.fullmap <- bind_rows(FamB.fullmap, temp) %>%
  group_by(LG) %>%
  arrange(POSITION) %>%
  ungroup()

# write output to file
write.table(FamB.fullmap, "results/FamB.fullmap-1", row.names = FALSE)

```


# Compare family maps after removing and re-mapping markers I

Load family maps and merge based on shared markers to compare sequence of shared markers in family A and B map.

```{r}

# load family map (with duplicate markers added back in)
FamA <- read.table("data/LINKMAP/FamA.fullmap",
                   stringsAsFactors = FALSE, sep = "", skip = 1,
                   col.names = c("LG_FamA","LOCUS", "POS_FamA"))

FamB <- read.table("data/LINKMAP/FamB.fullmap",
                   stringsAsFactors = FALSE, sep = "", skip = 1,
                   col.names = c("LG_FamB","LOCUS", "POS_FamB"))

temp <- inner_join(FamA, FamB, by = "LOCUS")

write.table(temp, "results/shared.markers-1",
            quote = FALSE, col.names = TRUE, row.names = FALSE)

```

Compare family maps. Identify problematic loci.

## Family A

```{r}

# change header names to Marker | Markertype | Genotype
FamA.onemap <- read.delim("data/LINKMAP/FamA.onemap",
                          header = FALSE, sep = "\t",
                          col.names = c("LOCUS", "LOCUSTYPE", "GENOTYPE"),
                          skip = 1, stringsAsFactors = FALSE)

# number of individuals in dataset
N <- read.delim("data/LINKMAP/FamA.onemap", header = FALSE,
                          col.names = c("N_IND", "N_LOCI"),
                          nrows = 1, stringsAsFactors = FALSE)

# load file with mapped markers and filter onemap file
mapped_markers <- read.table("results/FamA_allmarkers.txt", 
                             header = TRUE, stringsAsFactors = TRUE) %>%
    select(LOCUS) %>%
    transmute(LOCUS = paste(c("*"), LOCUS, sep = ""))

FamA.onemap.filtered <- semi_join(FamA.onemap, mapped_markers)

# remove flagged markers
FamA.onemap.filtered <- FamA.onemap.filtered %>%
    filter(!LOCUS %in% c("*dDocent_Contig_39676",
                         "*dDocent_Contig_34189",
                         "*dDocent_Contig_53933",
                         "*dDocent_Contig_20895",
                         "*dDocent_Contig_25692",
                         "*dDocent_Contig_9482",
                         "*dDocent_Contig_169742",
                         "*dDocent_Contig_42799",
                         "*dDocent_Contig_83367",
                         "*dDocent_Contig_38486",
                         "*dDocent_Contig_29327",
                         "*dDocent_Contig_84389",
                         "*dDocent_Contig_24719",
                         "*dDocent_Contig_91792",
                         "*dDocent_Contig_121713",
                         "*dDocent_Contig_12068",
                         "*dDocent_Contig_59623",
                         "*dDocent_Contig_41944",
                         "*dDocent_Contig_26467",
                         "*dDocent_Contig_30805",
                         "*dDocent_Contig_248473",
                         "*dDocent_Contig_244117",
                         "*dDocent_Contig_11546",
                         "*dDocent_Contig_35991",
                         "*dDocent_Contig_35624",
                         "*dDocent_Contig_200224",
                         "*dDocent_Contig_7091",
                         "*dDocent_Contig_2358",
                         "*dDocent_Contig_23388",
                         "*dDocent_Contig_186256",
                         "*dDocent_Contig_3817",
                         "*dDocent_Contig_97890",
                         "*dDocent_Contig_298306",
                         "*dDocent_Contig_67843",
                         "*dDocent_Contig_29969",
                         "*dDocent_Contig_15670",
                         "*dDocent_Contig_30792",
                         "*dDocent_Contig_75671",
                         "*dDocent_Contig_172925",
                         "*dDocent_Contig_144097",
                         "*dDocent_Contig_64257",
                         "*dDocent_Contig_12459",
                         "*dDocent_Contig_44930",
                         "*dDocent_Contig_990",
                         "*dDocent_Contig_50993",
                         "*dDocent_Contig_8255",
                         "*dDocent_Contig_20685",
                         "*dDocent_Contig_57614",
                         "*dDocent_Contig_3486",
                         "*dDocent_Contig_69118",
                         "*dDocent_Contig_51033",
                         "*dDocent_Contig_34901",
                         "*dDocent_Contig_1452",
                         "*dDocent_Contig_2847",
                         "*dDocent_Contig_6716",
                         "*dDocent_Contig_60349",
                         "*dDocent_Contig_77680",
                         "*dDocent_Contig_12222",
                         "*dDocent_Contig_62798",
                         "*dDocent_Contig_87392",
                         "*dDocent_Contig_39281",
                         "*dDocent_Contig_13012",
                         "*dDocent_Contig_44363",
                         "*dDocent_Contig_52542",
                         "*dDocent_Contig_123370",
                         "*dDocent_Contig_63085",
                         "*dDocent_Contig_138814",
                         "*dDocent_Contig_261231",
                         "*dDocent_Contig_47638",
                         "*dDocent_Contig_9761",
                         "*dDocent_Contig_4017",
                         "*dDocent_Contig_233860",
                         "*dDocent_Contig_24796",
                         "*dDocent_Contig_30828",
                         "*dDocent_Contig_52309",
                         "*dDocent_Contig_78689",
                         "*dDocent_Contig_62261",
                         "*dDocent_Contig_97106",
                         "*dDocent_Contig_1694",
                         "*dDocent_Contig_184316",
                         "*dDocent_Contig_483",
                         "*dDocent_Contig_596",
                         "*dDocent_Contig_13530",
                         "*dDocent_Contig_14522",
                         "*dDocent_Contig_50994",
                         "*dDocent_Contig_315477",
                         "*dDocent_Contig_730",
                         "*dDocent_Contig_107875",
                         "*dDocent_Contig_41408",
                         '*dDocent_Contig_213918',
                         "*dDocent_Contig_130476",
                         "*dDocent_Contig_129129",
                         "*dDocent_Contig_69979",
                         "*dDocent_Contig_224858",
                         "*dDocent_Contig_14784",
                         "*dDocent_Contig_17546",
                         "*dDocent_Contig_169675",
                         "*dDocent_Contig_26617",
                         "*dDocent_Contig_180953",
                         "*dDocent_Contig_16379",
                         "*dDocent_Contig_333713",
                         "*dDocent_Contig_52478",
                         "*dDocent_Contig_294562",
                         "*dDocent_Contig_17020",
                         "*dDocent_Contig_29858",
                         "*dDocent_Contig_6740",
                         "*dDocent_Contig_59297",
                         "*dDocent_Contig_3323",
                         "*dDocent_Contig_104837",
                         "*dDocent_Contig_305075",
                         "*dDocent_Contig_52003",
                         "*dDocent_Contig_29931",
                         "*dDocent_Contig_10018",
                         "*dDocent_Contig_65233",
                         "*dDocent_Contig_2867",
                         "*dDocent_Contig_42394",
                         "*dDocent_Contig_360",
                         "*dDocent_Contig_25067"))

# remove duplicate markers
temp <- FamA.onemap.filtered %>%
    distinct(GENOTYPE, .keep_all = TRUE)
    
# create dataframe of removed loci and corresponding loci
match_locA <- anti_join(FamA.onemap.filtered, temp) %>%
  rename(MATCHLOC = LOCUS) %>%
  select(MATCHLOC, GENOTYPE) %>%
  left_join(temp, by = "GENOTYPE") %>%
  select(LOCUS, MATCHLOC)

# write filtered input file for Onemap
# need to add extra line back in
con <- file("data/LINKMAP/FamAfilter.onemap", open = "wt")
writeLines(paste(N$N_IND, (nrow(temp))), con)
write.table(temp, con,
            append = FALSE, quote = FALSE, sep = " ",
            na = "NA", dec = ".",
            col.names = FALSE, row.names = FALSE)
close(con)

# import datafile
FamA.Map <- read.outcross(file="data/LINKMAP/FamAfilter.onemap")

# Summary of datafile
FamA.Map

# estimate recombination fraction between all pairs of markers
# creates object of class rf.2pts
FamA.Map.rf.L6.rf35 <- rf.2pts(FamA.Map, LOD = 6, max.rf = 0.35, verbose=TRUE)

# create sequence (list) of all markers to assign to LGs
all.markers <- make.seq(FamA.Map.rf.L6.rf35, "all")

# assign markers to linkage groups
LGs <- group(all.markers, LOD = 6, max.rf = 0.35)

# define map function
set.map.fun(typ = "haldane")

# create list of separate LGs
# (convert from class group to class sequence)
LG_list <- lapply(c(1:24), function(LG) {
  make.seq(LGs, LG)
})

# uses compare() to create a framework of most informative markers
# map remaining markers using try.seq()
# this step will take a few hours
LG_ordered <- lapply((LG_list), function(LG){
  order.seq(LG, n.init=6, THRES=3, touchdown=TRUE, tol=10e-2)
})

# create marker order with all markers
LG_final <- lapply((LG_ordered), function(LG){
  make.seq(LG, "force")
})

# group all final ordered LGs
FamA.map <- LG_final

# draw map for all LGs
draw.map(FamA.map, names=FALSE, grid=TRUE, cex.mrk=0.7)

# export map to file
write.map(FamA.map, "data/LINKMAP/FamA.fullmap")

FamA.fullmap <- read.table("data/LINKMAP/FamA.fullmap",
                           skip = 1, header = FALSE) %>%
  mutate(LOCUS = paste(c("*"), V2, sep = "")) %>%
  rename(LG = V1, POSITION = V3) %>%
  select(LOCUS, LG, POSITION)

# add mapping positions
temp <- left_join(match_locA, FamA.fullmap, by = "LOCUS") %>%
  select(MATCHLOC, LG, POSITION) %>%
  rename(LOCUS = MATCHLOC)

# add markers to final map
FamA.fullmap <- bind_rows(FamA.fullmap, temp) %>%
  group_by(LG) %>%
  arrange(POSITION) %>%
  ungroup()

# write output to file
write.table(FamA.fullmap, "results/FamA-2.fullmap", row.names = FALSE)

```

## Family B:

```{r}

# change header names to Marker | Markertype | Genotype
FamB.onemap <- read.delim("data/LINKMAP/FamB.onemap",
                          header = FALSE, sep = "\t",
                          col.names = c("LOCUS", "LOCUSTYPE", "GENOTYPE"),
                          skip = 1, stringsAsFactors = FALSE)

# number of individuals in dataset
N <- read.delim("data/LINKMAP/FamB.onemap", header = FALSE,
                          col.names = c("N_IND", "N_LOCI"),
                          nrows = 1, stringsAsFactors = FALSE)

# load file with mapped markers and filter onemap file
mapped_markers <- read.table("results/FamB_allmarkers.txt", 
                             header = TRUE, stringsAsFactors = TRUE) %>%
    select(LOCUS) %>%
    transmute(LOCUS = paste(c("*"), LOCUS, sep = ""))

FamB.onemap.filtered <- semi_join(FamB.onemap, mapped_markers)

# remove flagged markers
FamB.onemap.filtered <- FamB.onemap.filtered %>%
    filter(!LOCUS %in% c("*dDocent_Contig_30805",
                         "*dDocent_Contig_248473",
                         "*dDocent_Contig_244117",
                         "*dDocent_Contig_11546",
                         "*dDocent_Contig_35991",
                         "*dDocent_Contig_35624",
                         "*dDocent_Contig_200224",
                         "*dDocent_Contig_7091",
                         "*dDocent_Contig_2358",
                         "*dDocent_Contig_23388",
                         "*dDocent_Contig_186256",
                         "*dDocent_Contig_3817",
                         "*dDocent_Contig_97890",
                         "*dDocent_Contig_298306",
                         "*dDocent_Contig_67843",
                         "*dDocent_Contig_29969",
                         "*dDocent_Contig_15670",
                         "*dDocent_Contig_30792",
                         "*dDocent_Contig_75671",
                         "*dDocent_Contig_172925",
                         "*dDocent_Contig_144097",
                         "*dDocent_Contig_64257",
                         "*dDocent_Contig_12459",
                         "*dDocent_Contig_44930",
                         "*dDocent_Contig_990",
                         "*dDocent_Contig_50993",
                         "*dDocent_Contig_8255",
                         "*dDocent_Contig_20685",
                         "*dDocent_Contig_57614",
                         "*dDocent_Contig_3486",
                         "*dDocent_Contig_69118",
                         "*dDocent_Contig_51033",
                         "*dDocent_Contig_34901",
                         "*dDocent_Contig_1452",
                         "*dDocent_Contig_2847",
                         "*dDocent_Contig_6716",
                         "*dDocent_Contig_60349",
                         "*dDocent_Contig_77680",
                         "*dDocent_Contig_12222",
                         "*dDocent_Contig_62798",
                         "*dDocent_Contig_87392",
                         "*dDocent_Contig_39281",
                         "*dDocent_Contig_13012",
                         "*dDocent_Contig_44363",
                         "*dDocent_Contig_52542",
                         "*dDocent_Contig_123370",
                         "*dDocent_Contig_63085",
                         "*dDocent_Contig_138814",
                         "*dDocent_Contig_261231",
                         "*dDocent_Contig_47638",
                         "*dDocent_Contig_9761",
                         "*dDocent_Contig_4017",
                         "*dDocent_Contig_233860",
                         "*dDocent_Contig_24796",
                         "*dDocent_Contig_30828",
                         "*dDocent_Contig_52309",
                         "*dDocent_Contig_78689",
                         "*dDocent_Contig_62261",
                         "*dDocent_Contig_97106",
                         "*dDocent_Contig_1694",
                         "*dDocent_Contig_184316",
                         "*dDocent_Contig_483",
                         "*dDocent_Contig_596",
                         "*dDocent_Contig_13530",
                         "*dDocent_Contig_14522",
                         "*dDocent_Contig_50994",
                         "*dDocent_Contig_315477",
                         "*dDocent_Contig_730",
                         "*dDocent_Contig_107875",
                         "*dDocent_Contig_41408",
                         "*dDocent_Contig_213918",
                         "*dDocent_Contig_130476",
                         "*dDocent_Contig_129129",
                         "*dDocent_Contig_69979",
                         "*dDocent_Contig_224858",
                         "*dDocent_Contig_14784",
                         "*dDocent_Contig_17546",
                         "*dDocent_Contig_169675",
                         "*dDocent_Contig_26617",
                         "*dDocent_Contig_180953",
                         "*dDocent_Contig_16379",
                         "*dDocent_Contig_333713",
                         "*dDocent_Contig_52478",
                         "*dDocent_Contig_294562",
                         "*dDocent_Contig_17020",
                         "*dDocent_Contig_29858",
                         "*dDocent_Contig_6740",
                         "*dDocent_Contig_59297",
                         "*dDocent_Contig_3323",
                         "*dDocent_Contig_104837",
                         "*dDocent_Contig_305075",
                         "*dDocent_Contig_52003",
                         "*dDocent_Contig_29931",
                         "*dDocent_Contig_10018",
                         "*dDocent_Contig_65233",
                         "*dDocent_Contig_2867",
                         "*dDocent_Contig_42394",
                         "*dDocent_Contig_360",
                         "*dDocent_Contig_25067",
                         "*dDocent_Contig_39676",
                         "*dDocent_Contig_307589",
                         "*dDocent_Contig_23860",
                         "*dDocent_Contig_83810",
                         "*dDocent_Contig_11000",
                         "*dDocent_Contig_110843",
                         "*dDocent_Contig_32599",
                         "*dDocent_Contig_317880",
                         "*dDocent_Contig_325604",
                         "*dDocent_Contig_257",
                         "*dDocent_Contig_99894",
                         "*dDocent_Contig_10702",
                         "*dDocent_Contig_388381",
                         "*dDocent_Contig_236369",
                         "*dDocent_Contig_301685",
                         "*dDocent_Contig_30077",
                         "*dDocent_Contig_40834",
                         "*dDocent_Contig_4522",
                         "*dDocent_Contig_13137",
                         "*dDocent_Contig_182486",
                         "*dDocent_Contig_14563",
                         "*dDocent_Contig_53181",
                         "*dDocent_Contig_30202",
                         "*dDocent_Contig_34603",
                         "*dDocent_Contig_80211",
                         "*dDocent_Contig_147429",
                         "*dDocent_Contig_61940",
                         "*dDocent_Contig_14591",
                         "*dDocent_Contig_12218",
                         "*dDocent_Contig_21407",
                         "*dDocent_Contig_21399",
                         "*dDocent_Contig_89781",
                         "*dDocent_Contig_9840",
                         "*dDocent_Contig_31850",
                         "*dDocent_Contig_10808",
                         "*dDocent_Contig_360054",
                         "*dDocent_Contig_23809",
                         "*dDocent_Contig_36117",
                         "*dDocent_Contig_15730",
                         "*dDocent_Contig_390204",
                         "*dDocent_Contig_63932",
                         "*dDocent_Contig_48250",
                         "*dDocent_Contig_394960"))

# remove duplicate markers
temp <- FamB.onemap.filtered %>%
    distinct(GENOTYPE, .keep_all = TRUE)
    
# create dataframe of removed loci and corresponding loci
match_locB <- anti_join(FamB.onemap.filtered, temp) %>%
  rename(MATCHLOC = LOCUS) %>%
  select(MATCHLOC, GENOTYPE) %>%
  left_join(temp, by = "GENOTYPE") %>%
  select(LOCUS, MATCHLOC)

# write filtered input file for Onemap
# need to add extra line back in
con <- file("data/LINKMAP/FamBfilter.onemap", open = "wt")
writeLines(paste(N$N_IND, (nrow(temp))), con)
write.table(temp, con,
            append = FALSE, quote = FALSE, sep = " ",
            na = "NA", dec = ".",
            col.names = FALSE, row.names = FALSE)
close(con)

# import datafile
FamB.Map <- read.outcross(file="data/LINKMAP/FamBfilter.onemap")

# Summary of datafile
FamB.Map

# estimate recombination fraction between all pairs of markers
# creates object of class rf.2pts
FamB.Map.rf.L6.rf35 <- rf.2pts(FamB.Map, LOD = 6, max.rf = 0.35, verbose=TRUE)

# create sequence (list) of all markers to assign to LGs
all.markers <- make.seq(FamB.Map.rf.L6.rf35, "all")

# assign markers to linkage groups
LGs <- group(all.markers, LOD = 6, max.rf = 0.35)

# define map function
set.map.fun(typ = "haldane")

# create list of separate LGs
# (convert from class group to class sequence)
LG_list <- lapply(c(1:24), function(LG) {
  make.seq(LGs, LG)
})

# uses compare() to create a framework of most informative markers
# map remaining markers using try.seq()
# this step will take a few hours
LG_ordered <- lapply((LG_list), function(LG){
  order.seq(LG, n.init=6, THRES=3, touchdown=TRUE, tol=10e-2)
})

# create marker order with all markers
LG_final <- lapply((LG_ordered), function(LG){
  make.seq(LG, "force")
})

# group all final ordered LGs
FamB.map <- LG_final

# draw map for all LGs
draw.map(FamB.map, names=FALSE, grid=TRUE, cex.mrk=0.7)

# export map to file
write.map(FamB.map, "data/LINKMAP/FamB.fullmap")

FamB.fullmap <- read.table("data/LINKMAP/FamB.fullmap",
                           skip = 1, header = FALSE) %>%
  mutate(LOCUS = paste(c("*"), V2, sep = "")) %>%
  rename(LG = V1, POSITION = V3) %>%
  select(LOCUS, LG, POSITION)

# add mapping positions
temp <- left_join(match_locB, FamB.fullmap, by = "LOCUS") %>%
  select(MATCHLOC, LG, POSITION) %>%
  rename(LOCUS = MATCHLOC)

# add markers to final map
FamB.fullmap <- bind_rows(FamB.fullmap, temp) %>%
  group_by(LG) %>%
  arrange(POSITION) %>%
  ungroup()

# write output to file
write.table(FamB.fullmap, "results/FamB-2.fullmap", row.names = FALSE)

```

# Compare family maps after removing and re-mapping markers II

Load family maps and merge based on shared markers to compare sequence of shared markers in family A and B map.

```{r}

# load family map (with duplicate markers added back in)
FamA <- read.table("results/FamA-2.fullmap",
                   stringsAsFactors = FALSE, sep = "", skip = 1,
                   col.names = c("LOCUS", "LG_FamA", "POS_FamA"))

FamB <- read.table("results/FamB-2.fullmap",
                   stringsAsFactors = FALSE, sep = "", skip = 1,
                   col.names = c("LOCUS", "LG_FamB", "POS_FamB"))

temp <- inner_join(FamA, FamB, by = "LOCUS")

write.table(temp, "results/shared.markers-2",
            quote = FALSE, col.names = TRUE, row.names = FALSE)

```

Compare family maps. Identify problematic loci.

## Family A

```{r}

# change header names to Marker | Markertype | Genotype
FamA.onemap <- read.delim("data/LINKMAP/FamA.onemap",
                          header = FALSE, sep = "\t",
                          col.names = c("LOCUS", "LOCUSTYPE", "GENOTYPE"),
                          skip = 1, stringsAsFactors = FALSE)

# number of individuals in dataset
N <- read.delim("data/LINKMAP/FamA.onemap", header = FALSE,
                          col.names = c("N_IND", "N_LOCI"),
                          nrows = 1, stringsAsFactors = FALSE)

# load file with mapped markers and filter onemap file
mapped_markers <- read.table("results/FamA_allmarkers.txt", 
                             header = TRUE, stringsAsFactors = TRUE) %>%
    select(LOCUS) %>%
    transmute(LOCUS = paste(c("*"), LOCUS, sep = ""))

FamA.onemap.filtered <- semi_join(FamA.onemap, mapped_markers)

# remove flagged markers
FamA.onemap.filtered <- FamA.onemap.filtered %>%
    filter(!LOCUS %in% c("*dDocent_Contig_39676",
                         "*dDocent_Contig_34189",
                         "*dDocent_Contig_53933",
                         "*dDocent_Contig_20895",
                         "*dDocent_Contig_25692",
                         "*dDocent_Contig_9482",
                         "*dDocent_Contig_169742",
                         "*dDocent_Contig_42799",
                         "*dDocent_Contig_83367",
                         "*dDocent_Contig_38486",
                         "*dDocent_Contig_29327",
                         "*dDocent_Contig_84389",
                         "*dDocent_Contig_24719",
                         "*dDocent_Contig_91792",
                         "*dDocent_Contig_121713",
                         "*dDocent_Contig_12068",
                         "*dDocent_Contig_59623",
                         "*dDocent_Contig_41944",
                         "*dDocent_Contig_26467",
                         "*dDocent_Contig_30805",
                         "*dDocent_Contig_248473",
                         "*dDocent_Contig_244117",
                         "*dDocent_Contig_11546",
                         "*dDocent_Contig_35991",
                         "*dDocent_Contig_35624",
                         "*dDocent_Contig_200224",
                         "*dDocent_Contig_7091",
                         "*dDocent_Contig_2358",
                         "*dDocent_Contig_23388",
                         "*dDocent_Contig_186256",
                         "*dDocent_Contig_3817",
                         "*dDocent_Contig_97890",
                         "*dDocent_Contig_298306",
                         "*dDocent_Contig_67843",
                         "*dDocent_Contig_29969",
                         "*dDocent_Contig_15670",
                         "*dDocent_Contig_30792",
                         "*dDocent_Contig_75671",
                         "*dDocent_Contig_172925",
                         "*dDocent_Contig_144097",
                         "*dDocent_Contig_64257",
                         "*dDocent_Contig_12459",
                         "*dDocent_Contig_44930",
                         "*dDocent_Contig_990",
                         "*dDocent_Contig_50993",
                         "*dDocent_Contig_8255",
                         "*dDocent_Contig_20685",
                         "*dDocent_Contig_57614",
                         "*dDocent_Contig_3486",
                         "*dDocent_Contig_69118",
                         "*dDocent_Contig_51033",
                         "*dDocent_Contig_34901",
                         "*dDocent_Contig_1452",
                         "*dDocent_Contig_2847",
                         "*dDocent_Contig_6716",
                         "*dDocent_Contig_60349",
                         "*dDocent_Contig_77680",
                         "*dDocent_Contig_12222",
                         "*dDocent_Contig_62798",
                         "*dDocent_Contig_87392",
                         "*dDocent_Contig_39281",
                         "*dDocent_Contig_13012",
                         "*dDocent_Contig_44363",
                         "*dDocent_Contig_52542",
                         "*dDocent_Contig_123370",
                         "*dDocent_Contig_63085",
                         "*dDocent_Contig_138814",
                         "*dDocent_Contig_261231",
                         "*dDocent_Contig_47638",
                         "*dDocent_Contig_9761",
                         "*dDocent_Contig_4017",
                         "*dDocent_Contig_233860",
                         "*dDocent_Contig_24796",
                         "*dDocent_Contig_30828",
                         "*dDocent_Contig_52309",
                         "*dDocent_Contig_78689",
                         "*dDocent_Contig_62261",
                         "*dDocent_Contig_97106",
                         "*dDocent_Contig_1694",
                         "*dDocent_Contig_184316",
                         "*dDocent_Contig_483",
                         "*dDocent_Contig_596",
                         "*dDocent_Contig_13530",
                         "*dDocent_Contig_14522",
                         "*dDocent_Contig_50994",
                         "*dDocent_Contig_315477",
                         "*dDocent_Contig_730",
                         "*dDocent_Contig_107875",
                         "*dDocent_Contig_41408",
                         '*dDocent_Contig_213918',
                         "*dDocent_Contig_130476",
                         "*dDocent_Contig_129129",
                         "*dDocent_Contig_69979",
                         "*dDocent_Contig_224858",
                         "*dDocent_Contig_14784",
                         "*dDocent_Contig_17546",
                         "*dDocent_Contig_169675",
                         "*dDocent_Contig_26617",
                         "*dDocent_Contig_180953",
                         "*dDocent_Contig_16379",
                         "*dDocent_Contig_333713",
                         "*dDocent_Contig_52478",
                         "*dDocent_Contig_294562",
                         "*dDocent_Contig_17020",
                         "*dDocent_Contig_29858",
                         "*dDocent_Contig_6740",
                         "*dDocent_Contig_59297",
                         "*dDocent_Contig_3323",
                         "*dDocent_Contig_104837",
                         "*dDocent_Contig_305075",
                         "*dDocent_Contig_52003",
                         "*dDocent_Contig_29931",
                         "*dDocent_Contig_10018",
                         "*dDocent_Contig_65233",
                         "*dDocent_Contig_2867",
                         "*dDocent_Contig_42394",
                         "*dDocent_Contig_360",
                         "*dDocent_Contig_25067",
                         "*dDocent_Contig_200617",
                         "*dDocent_Contig_46589",
                         "*dDocent_Contig_117120",
                         "*dDocent_Contig_27435",
                         "*dDocent_Contig_19718",
                         "*dDocent_Contig_9989",
                         "*dDocent_Contig_35456",
                         "*dDocent_Contig_4468"))

# remove duplicate markers
temp <- FamA.onemap.filtered %>%
    distinct(GENOTYPE, .keep_all = TRUE)
    
# create dataframe of removed loci and corresponding loci
match_locA <- anti_join(FamA.onemap.filtered, temp) %>%
  rename(MATCHLOC = LOCUS) %>%
  select(MATCHLOC, GENOTYPE) %>%
  left_join(temp, by = "GENOTYPE") %>%
  select(LOCUS, MATCHLOC)

# write filtered input file for Onemap
# need to add extra line back in
con <- file("data/LINKMAP/FamAfilter.onemap", open = "wt")
writeLines(paste(N$N_IND, (nrow(temp))), con)
write.table(temp, con,
            append = FALSE, quote = FALSE, sep = " ",
            na = "NA", dec = ".",
            col.names = FALSE, row.names = FALSE)
close(con)

# import datafile
FamA.Map <- read.outcross(file="data/LINKMAP/FamAfilter.onemap")

# Summary of datafile
FamA.Map

# estimate recombination fraction between all pairs of markers
# creates object of class rf.2pts
FamA.Map.rf.L6.rf35 <- rf.2pts(FamA.Map, LOD = 6, max.rf = 0.35, verbose=TRUE)

# create sequence (list) of all markers to assign to LGs
all.markers <- make.seq(FamA.Map.rf.L6.rf35, "all")

# assign markers to linkage groups
LGs <- group(all.markers, LOD = 6, max.rf = 0.35)

# define map function
set.map.fun(typ = "haldane")

# create list of separate LGs
# (convert from class group to class sequence)
LG_list <- lapply(c(1:24), function(LG) {
  make.seq(LGs, LG)
})

# uses compare() to create a framework of most informative markers
# map remaining markers using try.seq()
# this step will take a few hours
LG_ordered <- lapply((LG_list), function(LG){
  order.seq(LG, n.init=6, THRES=3, touchdown=TRUE, tol=10e-2)
})

# create marker order with all markers
LG_final <- lapply((LG_ordered), function(LG){
  make.seq(LG, "force")
})

# group all final ordered LGs
FamA.map <- LG_final

# draw map for all LGs
draw.map(FamA.map, names=FALSE, grid=TRUE, cex.mrk=0.7)

# export map to file
write.map(FamA.map, "data/LINKMAP/FamA.fullmap")

FamA.fullmap <- read.table("data/LINKMAP/FamA.fullmap",
                           skip = 1, header = FALSE) %>%
  mutate(LOCUS = paste(c("*"), V2, sep = "")) %>%
  rename(LG = V1, POSITION = V3) %>%
  select(LOCUS, LG, POSITION)

# add mapping positions
temp <- left_join(match_locA, FamA.fullmap, by = "LOCUS") %>%
  select(MATCHLOC, LG, POSITION) %>%
  rename(LOCUS = MATCHLOC)

# add markers to final map
FamA.fullmap <- bind_rows(FamA.fullmap, temp) %>%
  group_by(LG) %>%
  arrange(POSITION) %>%
  ungroup()

# write output to file
write.table(FamA.fullmap, "results/FamA-3.fullmap", row.names = FALSE)

```

## Family B:

```{r}

# change header names to Marker | Markertype | Genotype
FamB.onemap <- read.delim("data/LINKMAP/FamB.onemap",
                          header = FALSE, sep = "\t",
                          col.names = c("LOCUS", "LOCUSTYPE", "GENOTYPE"),
                          skip = 1, stringsAsFactors = FALSE)

# number of individuals in dataset
N <- read.delim("data/LINKMAP/FamB.onemap", header = FALSE,
                          col.names = c("N_IND", "N_LOCI"),
                          nrows = 1, stringsAsFactors = FALSE)

# load file with mapped markers and filter onemap file
mapped_markers <- read.table("results/FamB_allmarkers.txt", 
                             header = TRUE, stringsAsFactors = TRUE) %>%
    select(LOCUS) %>%
    transmute(LOCUS = paste(c("*"), LOCUS, sep = ""))

FamB.onemap.filtered <- semi_join(FamB.onemap, mapped_markers)

# remove flagged markers
FamB.onemap.filtered <- FamB.onemap.filtered %>%
    filter(!LOCUS %in% c("*dDocent_Contig_30805",
                         "*dDocent_Contig_248473",
                         "*dDocent_Contig_244117",
                         "*dDocent_Contig_11546",
                         "*dDocent_Contig_35991",
                         "*dDocent_Contig_35624",
                         "*dDocent_Contig_200224",
                         "*dDocent_Contig_7091",
                         "*dDocent_Contig_2358",
                         "*dDocent_Contig_23388",
                         "*dDocent_Contig_186256",
                         "*dDocent_Contig_3817",
                         "*dDocent_Contig_97890",
                         "*dDocent_Contig_298306",
                         "*dDocent_Contig_67843",
                         "*dDocent_Contig_29969",
                         "*dDocent_Contig_15670",
                         "*dDocent_Contig_30792",
                         "*dDocent_Contig_75671",
                         "*dDocent_Contig_172925",
                         "*dDocent_Contig_144097",
                         "*dDocent_Contig_64257",
                         "*dDocent_Contig_12459",
                         "*dDocent_Contig_44930",
                         "*dDocent_Contig_990",
                         "*dDocent_Contig_50993",
                         "*dDocent_Contig_8255",
                         "*dDocent_Contig_20685",
                         "*dDocent_Contig_57614",
                         "*dDocent_Contig_3486",
                         "*dDocent_Contig_69118",
                         "*dDocent_Contig_51033",
                         "*dDocent_Contig_34901",
                         "*dDocent_Contig_1452",
                         "*dDocent_Contig_2847",
                         "*dDocent_Contig_6716",
                         "*dDocent_Contig_60349",
                         "*dDocent_Contig_77680",
                         "*dDocent_Contig_12222",
                         "*dDocent_Contig_62798",
                         "*dDocent_Contig_87392",
                         "*dDocent_Contig_39281",
                         "*dDocent_Contig_13012",
                         "*dDocent_Contig_44363",
                         "*dDocent_Contig_52542",
                         "*dDocent_Contig_123370",
                         "*dDocent_Contig_63085",
                         "*dDocent_Contig_138814",
                         "*dDocent_Contig_261231",
                         "*dDocent_Contig_47638",
                         "*dDocent_Contig_9761",
                         "*dDocent_Contig_4017",
                         "*dDocent_Contig_233860",
                         "*dDocent_Contig_24796",
                         "*dDocent_Contig_30828",
                         "*dDocent_Contig_52309",
                         "*dDocent_Contig_78689",
                         "*dDocent_Contig_62261",
                         "*dDocent_Contig_97106",
                         "*dDocent_Contig_1694",
                         "*dDocent_Contig_184316",
                         "*dDocent_Contig_483",
                         "*dDocent_Contig_596",
                         "*dDocent_Contig_13530",
                         "*dDocent_Contig_14522",
                         "*dDocent_Contig_50994",
                         "*dDocent_Contig_315477",
                         "*dDocent_Contig_730",
                         "*dDocent_Contig_107875",
                         "*dDocent_Contig_41408",
                         "*dDocent_Contig_213918",
                         "*dDocent_Contig_130476",
                         "*dDocent_Contig_129129",
                         "*dDocent_Contig_69979",
                         "*dDocent_Contig_224858",
                         "*dDocent_Contig_14784",
                         "*dDocent_Contig_17546",
                         "*dDocent_Contig_169675",
                         "*dDocent_Contig_26617",
                         "*dDocent_Contig_180953",
                         "*dDocent_Contig_16379",
                         "*dDocent_Contig_333713",
                         "*dDocent_Contig_52478",
                         "*dDocent_Contig_294562",
                         "*dDocent_Contig_17020",
                         "*dDocent_Contig_29858",
                         "*dDocent_Contig_6740",
                         "*dDocent_Contig_59297",
                         "*dDocent_Contig_3323",
                         "*dDocent_Contig_104837",
                         "*dDocent_Contig_305075",
                         "*dDocent_Contig_52003",
                         "*dDocent_Contig_29931",
                         "*dDocent_Contig_10018",
                         "*dDocent_Contig_65233",
                         "*dDocent_Contig_2867",
                         "*dDocent_Contig_42394",
                         "*dDocent_Contig_360",
                         "*dDocent_Contig_25067",
                         "*dDocent_Contig_39676",
                         "*dDocent_Contig_307589",
                         "*dDocent_Contig_23860",
                         "*dDocent_Contig_83810",
                         "*dDocent_Contig_11000",
                         "*dDocent_Contig_110843",
                         "*dDocent_Contig_32599",
                         "*dDocent_Contig_317880",
                         "*dDocent_Contig_325604",
                         "*dDocent_Contig_257",
                         "*dDocent_Contig_99894",
                         "*dDocent_Contig_10702",
                         "*dDocent_Contig_388381",
                         "*dDocent_Contig_236369",
                         "*dDocent_Contig_301685",
                         "*dDocent_Contig_30077",
                         "*dDocent_Contig_40834",
                         "*dDocent_Contig_4522",
                         "*dDocent_Contig_13137",
                         "*dDocent_Contig_182486",
                         "*dDocent_Contig_14563",
                         "*dDocent_Contig_53181",
                         "*dDocent_Contig_30202",
                         "*dDocent_Contig_34603",
                         "*dDocent_Contig_80211",
                         "*dDocent_Contig_147429",
                         "*dDocent_Contig_61940",
                         "*dDocent_Contig_14591",
                         "*dDocent_Contig_12218",
                         "*dDocent_Contig_21407",
                         "*dDocent_Contig_21399",
                         "*dDocent_Contig_89781",
                         "*dDocent_Contig_9840",
                         "*dDocent_Contig_31850",
                         "*dDocent_Contig_10808",
                         "*dDocent_Contig_360054",
                         "*dDocent_Contig_23809",
                         "*dDocent_Contig_36117",
                         "*dDocent_Contig_15730",
                         "*dDocent_Contig_390204",
                         "*dDocent_Contig_63932",
                         "*dDocent_Contig_48250",
                         "*dDocent_Contig_394960",
                         "*dDocent_Contig_23004",
                         "*dDocent_Contig_2987",
                         "*dDocent_Contig_17774",
                         "*dDocent_Contig_213871",
                         "*dDocent_Contig_12852",
                         "*dDocent_Contig_3932",
                         "*dDocent_Contig_21467",
                         "*dDocent_Contig_131586",
                         "*dDocent_Contig_3278",
                         "*dDocent_Contig_259416",
                         "*dDocent_Contig_35267",
                         "*dDocent_Contig_214291",
                         "*dDocent_Contig_6178",
                         "*dDocent_Contig_365182",
                         "*dDocent_Contig_122254",
                         "*dDocent_Contig_42515",
                         "*dDocent_Contig_255676",
                         "*dDocent_Contig_21480",
                         "*dDocent_Contig_30292"))

# remove duplicate markers
temp <- FamB.onemap.filtered %>%
    distinct(GENOTYPE, .keep_all = TRUE)
    
# create dataframe of removed loci and corresponding loci
match_locB <- anti_join(FamB.onemap.filtered, temp) %>%
  rename(MATCHLOC = LOCUS) %>%
  select(MATCHLOC, GENOTYPE) %>%
  left_join(temp, by = "GENOTYPE") %>%
  select(LOCUS, MATCHLOC)

# write filtered input file for Onemap
# need to add extra line back in
con <- file("data/LINKMAP/FamBfilter.onemap", open = "wt")
writeLines(paste(N$N_IND, (nrow(temp))), con)
write.table(temp, con,
            append = FALSE, quote = FALSE, sep = " ",
            na = "NA", dec = ".",
            col.names = FALSE, row.names = FALSE)
close(con)

# import datafile
FamB.Map <- read.outcross(file="data/LINKMAP/FamBfilter.onemap")

# Summary of datafile
FamB.Map

# estimate recombination fraction between all pairs of markers
# creates object of class rf.2pts
FamB.Map.rf.L6.rf35 <- rf.2pts(FamB.Map, LOD = 6, max.rf = 0.35, verbose=TRUE)

# create sequence (list) of all markers to assign to LGs
all.markers <- make.seq(FamB.Map.rf.L6.rf35, "all")

# assign markers to linkage groups
LGs <- group(all.markers, LOD = 6, max.rf = 0.35)

# define map function
set.map.fun(typ = "haldane")

# create list of separate LGs
# (convert from class group to class sequence)
LG_list <- lapply(c(1:24), function(LG) {
  make.seq(LGs, LG)
})

# uses compare() to create a framework of most informative markers
# map remaining markers using try.seq()
# this step will take a few hours
LG_ordered <- lapply((LG_list), function(LG){
  order.seq(LG, n.init=6, THRES=3, touchdown=TRUE, tol=10e-2)
})

# create marker order with all markers
LG_final <- lapply((LG_ordered), function(LG){
  make.seq(LG, "force")
})

# group all final ordered LGs
FamB.map <- LG_final

# draw map for all LGs
draw.map(FamB.map, names=FALSE, grid=TRUE, cex.mrk=0.7)

# export map to file
write.map(FamB.map, "data/LINKMAP/FamB.fullmap")

FamB.fullmap <- read.table("data/LINKMAP/FamB.fullmap",
                           skip = 1, header = FALSE) %>%
  mutate(LOCUS = paste(c("*"), V2, sep = "")) %>%
  rename(LG = V1, POSITION = V3) %>%
  select(LOCUS, LG, POSITION)

# add mapping positions
temp <- left_join(match_locB, FamB.fullmap, by = "LOCUS") %>%
  select(MATCHLOC, LG, POSITION) %>%
  rename(LOCUS = MATCHLOC)

# add markers to final map
FamB.fullmap <- bind_rows(FamB.fullmap, temp) %>%
  group_by(LG) %>%
  arrange(POSITION) %>%
  ungroup()

# write output to file
write.table(FamB.fullmap, "results/FamB-3.fullmap", row.names = FALSE)

```

# Compare family maps after removing and re-mapping markers III

Compare maps afte re-mapping.

```{r}

# load family map (with duplicate markers added back in)
FamA <- read.table("results/FamA-3.fullmap",
                   stringsAsFactors = FALSE, sep = "", skip = 1,
                   col.names = c("LOCUS", "LG_FamA","POS_FamA"))

FamB <- read.table("results/FamB-3.fullmap",
                   stringsAsFactors = FALSE, sep = "", skip = 1,
                   col.names = c("LOCUS", "LG_FamB","POS_FamB"))

temp <- inner_join(FamA, FamB, by = "LOCUS")

write.table(temp, "results/shared.markers-3",
            quote = FALSE, col.names = TRUE, row.names = FALSE)

```



## Family B:

```{r}

# change header names to Marker | Markertype | Genotype
FamB.onemap <- read.delim("data/LINKMAP/FamB.onemap",
                          header = FALSE, sep = "\t",
                          col.names = c("LOCUS", "LOCUSTYPE", "GENOTYPE"),
                          skip = 1, stringsAsFactors = FALSE)

# number of individuals in dataset
N <- read.delim("data/LINKMAP/FamB.onemap", header = FALSE,
                          col.names = c("N_IND", "N_LOCI"),
                          nrows = 1, stringsAsFactors = FALSE)

# load file with mapped markers and filter onemap file
mapped_markers <- read.table("results/FamB_allmarkers.txt", 
                             header = TRUE, stringsAsFactors = TRUE) %>%
    select(LOCUS) %>%
    transmute(LOCUS = paste(c("*"), LOCUS, sep = ""))

FamB.onemap.filtered <- semi_join(FamB.onemap, mapped_markers)

# remove flagged markers
FamB.onemap.filtered <- FamB.onemap.filtered %>%
    filter(!LOCUS %in% c("*dDocent_Contig_30805",
                         "*dDocent_Contig_248473",
                         "*dDocent_Contig_244117",
                         "*dDocent_Contig_11546",
                         "*dDocent_Contig_35991",
                         "*dDocent_Contig_35624",
                         "*dDocent_Contig_200224",
                         "*dDocent_Contig_7091",
                         "*dDocent_Contig_2358",
                         "*dDocent_Contig_23388",
                         "*dDocent_Contig_186256",
                         "*dDocent_Contig_3817",
                         "*dDocent_Contig_97890",
                         "*dDocent_Contig_298306",
                         "*dDocent_Contig_67843",
                         "*dDocent_Contig_29969",
                         "*dDocent_Contig_15670",
                         "*dDocent_Contig_30792",
                         "*dDocent_Contig_75671",
                         "*dDocent_Contig_172925",
                         "*dDocent_Contig_144097",
                         "*dDocent_Contig_64257",
                         "*dDocent_Contig_12459",
                         "*dDocent_Contig_44930",
                         "*dDocent_Contig_990",
                         "*dDocent_Contig_50993",
                         "*dDocent_Contig_8255",
                         "*dDocent_Contig_20685",
                         "*dDocent_Contig_57614",
                         "*dDocent_Contig_3486",
                         "*dDocent_Contig_69118",
                         "*dDocent_Contig_51033",
                         "*dDocent_Contig_34901",
                         "*dDocent_Contig_1452",
                         "*dDocent_Contig_2847",
                         "*dDocent_Contig_6716",
                         "*dDocent_Contig_60349",
                         "*dDocent_Contig_77680",
                         "*dDocent_Contig_12222",
                         "*dDocent_Contig_62798",
                         "*dDocent_Contig_87392",
                         "*dDocent_Contig_39281",
                         "*dDocent_Contig_13012",
                         "*dDocent_Contig_44363",
                         "*dDocent_Contig_52542",
                         "*dDocent_Contig_123370",
                         "*dDocent_Contig_63085",
                         "*dDocent_Contig_138814",
                         "*dDocent_Contig_261231",
                         "*dDocent_Contig_47638",
                         "*dDocent_Contig_9761",
                         "*dDocent_Contig_4017",
                         "*dDocent_Contig_233860",
                         "*dDocent_Contig_24796",
                         "*dDocent_Contig_30828",
                         "*dDocent_Contig_52309",
                         "*dDocent_Contig_78689",
                         "*dDocent_Contig_62261",
                         "*dDocent_Contig_97106",
                         "*dDocent_Contig_1694",
                         "*dDocent_Contig_184316",
                         "*dDocent_Contig_483",
                         "*dDocent_Contig_596",
                         "*dDocent_Contig_13530",
                         "*dDocent_Contig_14522",
                         "*dDocent_Contig_50994",
                         "*dDocent_Contig_315477",
                         "*dDocent_Contig_730",
                         "*dDocent_Contig_107875",
                         "*dDocent_Contig_41408",
                         "*dDocent_Contig_213918",
                         "*dDocent_Contig_130476",
                         "*dDocent_Contig_129129",
                         "*dDocent_Contig_69979",
                         "*dDocent_Contig_224858",
                         "*dDocent_Contig_14784",
                         "*dDocent_Contig_17546",
                         "*dDocent_Contig_169675",
                         "*dDocent_Contig_26617",
                         "*dDocent_Contig_180953",
                         "*dDocent_Contig_16379",
                         "*dDocent_Contig_333713",
                         "*dDocent_Contig_52478",
                         "*dDocent_Contig_294562",
                         "*dDocent_Contig_17020",
                         "*dDocent_Contig_29858",
                         "*dDocent_Contig_6740",
                         "*dDocent_Contig_59297",
                         "*dDocent_Contig_3323",
                         "*dDocent_Contig_104837",
                         "*dDocent_Contig_305075",
                         "*dDocent_Contig_52003",
                         "*dDocent_Contig_29931",
                         "*dDocent_Contig_10018",
                         "*dDocent_Contig_65233",
                         "*dDocent_Contig_2867",
                         "*dDocent_Contig_42394",
                         "*dDocent_Contig_360",
                         "*dDocent_Contig_25067",
                         "*dDocent_Contig_39676",
                         "*dDocent_Contig_307589",
                         "*dDocent_Contig_23860",
                         "*dDocent_Contig_83810",
                         "*dDocent_Contig_11000",
                         "*dDocent_Contig_110843",
                         "*dDocent_Contig_32599",
                         "*dDocent_Contig_317880",
                         "*dDocent_Contig_325604",
                         "*dDocent_Contig_257",
                         "*dDocent_Contig_99894",
                         "*dDocent_Contig_10702",
                         "*dDocent_Contig_388381",
                         "*dDocent_Contig_236369",
                         "*dDocent_Contig_301685",
                         "*dDocent_Contig_30077",
                         "*dDocent_Contig_40834",
                         "*dDocent_Contig_4522",
                         "*dDocent_Contig_13137",
                         "*dDocent_Contig_182486",
                         "*dDocent_Contig_14563",
                         "*dDocent_Contig_53181",
                         "*dDocent_Contig_30202",
                         "*dDocent_Contig_34603",
                         "*dDocent_Contig_80211",
                         "*dDocent_Contig_147429",
                         "*dDocent_Contig_61940",
                         "*dDocent_Contig_14591",
                         "*dDocent_Contig_12218",
                         "*dDocent_Contig_21407",
                         "*dDocent_Contig_21399",
                         "*dDocent_Contig_89781",
                         "*dDocent_Contig_9840",
                         "*dDocent_Contig_31850",
                         "*dDocent_Contig_10808",
                         "*dDocent_Contig_360054",
                         "*dDocent_Contig_23809",
                         "*dDocent_Contig_36117",
                         "*dDocent_Contig_15730",
                         "*dDocent_Contig_390204",
                         "*dDocent_Contig_63932",
                         "*dDocent_Contig_48250",
                         "*dDocent_Contig_394960",
                         "*dDocent_Contig_23004",
                         "*dDocent_Contig_2987",
                         "*dDocent_Contig_17774",
                         "*dDocent_Contig_213871",
                         "*dDocent_Contig_12852",
                         "*dDocent_Contig_3932",
                         "*dDocent_Contig_21467",
                         "*dDocent_Contig_131586",
                         "*dDocent_Contig_3278",
                         "*dDocent_Contig_259416",
                         "*dDocent_Contig_35267",
                         "*dDocent_Contig_214291",
                         "*dDocent_Contig_6178",
                         "*dDocent_Contig_365182",
                         "*dDocent_Contig_122254",
                         "*dDocent_Contig_42515",
                         "*dDocent_Contig_255676",
                         "*dDocent_Contig_21480",
                         "*dDocent_Contig_30292",
                         "*dDocent_Contig_77513",
                         "*dDocent_Contig_39037",
                         "*dDocent_Contig_26999",
                         "*dDocent_Contig_131658",
                         "*dDocent_Contig_178968",
                         "*dDocent_Contig_132810",
                         "*dDocent_Contig_14159",
                         "*dDocent_Contig_70156",
                         "*dDocent_Contig_98489",
                         "*dDocent_Contig_31046",
                         "*dDocent_Contig_15848",
                         "*dDocent_Contig_164687",
                         "*dDocent_Contig_2755",
                         "*dDocent_Contig_14160",
                         "*dDocent_Contig_63373",
                         "*dDocent_Contig_123246",
                         "*dDocent_Contig_3903"))

# remove duplicate markers
temp <- FamB.onemap.filtered %>%
    distinct(GENOTYPE, .keep_all = TRUE)
    
# create dataframe of removed loci and corresponding loci
match_locB <- anti_join(FamB.onemap.filtered, temp) %>%
  rename(MATCHLOC = LOCUS) %>%
  select(MATCHLOC, GENOTYPE) %>%
  left_join(temp, by = "GENOTYPE") %>%
  select(LOCUS, MATCHLOC)

# write filtered input file for Onemap
# need to add extra line back in
con <- file("data/LINKMAP/FamBfilter.onemap", open = "wt")
writeLines(paste(N$N_IND, (nrow(temp))), con)
write.table(temp, con,
            append = FALSE, quote = FALSE, sep = " ",
            na = "NA", dec = ".",
            col.names = FALSE, row.names = FALSE)
close(con)

# import datafile
FamB.Map <- read.outcross(file="data/LINKMAP/FamBfilter.onemap")

# Summary of datafile
FamB.Map

# estimate recombination fraction between all pairs of markers
# creates object of class rf.2pts
FamB.Map.rf.L6.rf35 <- rf.2pts(FamB.Map, LOD = 6, max.rf = 0.35, verbose=TRUE)

# create sequence (list) of all markers to assign to LGs
all.markers <- make.seq(FamB.Map.rf.L6.rf35, "all")

# assign markers to linkage groups
LGs <- group(all.markers, LOD = 6, max.rf = 0.35)

# define map function
set.map.fun(typ = "haldane")

# create list of separate LGs
# (convert from class group to class sequence)
LG_list <- lapply(c(1:24), function(LG) {
  make.seq(LGs, LG)
})

# uses compare() to create a framework of most informative markers
# map remaining markers using try.seq()
# this step will take a few hours
LG_ordered <- lapply((LG_list), function(LG){
  order.seq(LG, n.init=6, THRES=3, touchdown=TRUE, tol=10e-2)
})

# create marker order with all markers
LG_final <- lapply((LG_ordered), function(LG){
  make.seq(LG, "force")
})

# group all final ordered LGs
FamB.map <- LG_final

# draw map for all LGs
draw.map(FamB.map, names=FALSE, grid=TRUE, cex.mrk=0.7)

# export map to file
write.map(FamB.map, "data/LINKMAP/FamB.fullmap")

FamB.fullmap <- read.table("data/LINKMAP/FamB.fullmap",
                           skip = 1, header = FALSE) %>%
  mutate(LOCUS = paste(c("*"), V2, sep = "")) %>%
  rename(LG = V1, POSITION = V3) %>%
  select(LOCUS, LG, POSITION)

# add mapping positions
temp <- left_join(match_locB, FamB.fullmap, by = "LOCUS") %>%
  select(MATCHLOC, LG, POSITION) %>%
  rename(LOCUS = MATCHLOC)

# add markers to final map
FamB.fullmap <- bind_rows(FamB.fullmap, temp) %>%
  group_by(LG) %>%
  arrange(POSITION) %>%
  ungroup()

# write output to file
write.table(FamB.fullmap, "results/FamB-4.fullmap", row.names = FALSE)

```


```{r}

# load family map (with duplicate markers added back in)
FamA <- read.table("results/FamA-3.fullmap",
                   stringsAsFactors = FALSE, sep = "", skip = 1,
                   col.names = c("LOCUS", "LG_FamA","POS_FamA"))

FamB <- read.table("results/FamB-4.fullmap",
                   stringsAsFactors = FALSE, sep = "", skip = 1,
                   col.names = c("LOCUS", "LG_FamB","POS_FamB"))

temp <- inner_join(FamA, FamB, by = "LOCUS")

write.table(temp, "results/shared.markers-4",
            quote = FALSE, col.names = TRUE, row.names = FALSE)

```

# Merge Family maps

## Load files

Load family maps

```{r}

FamA <- read.delim("results/FamA_family.map", 
                   stringsAsFactors = FALSE, sep = "", header = TRUE)

FamB <- read.delim("results/FamB_family.map", 
                   header = TRUE, stringsAsFactors = FALSE, sep = "")

temp <- bind_rows(FamA, FamB)

all_markers <- unique(temp$LOCUS)

```

Each map consists of 24 linkage groups and a total of `r nrow(FamA)` and `r nrow(FamB)` markers for Family A and B, respectively with `r length(temp$LOCUS)-length(all_markers)` markers overlapping between the two family maps.

Use LPmerge to create consensus linkage map by merging linkage maps from Family A and B. LPmerge uses linear programming to minimize the mean absolute error between the consensus and family linkage maps; preserve ordering of the markers in the linkage map using linear inequality constraints. When marker order is inconsisten between maps being merged a minimum set of ordinal constraints is delected to resolve the conflicts.

## Merge maps for LG1:

```{r}

LG1_A <- FamA %>%
    filter(LG == 1) %>%
    select(LOCUS, POSITION)

LG1_B <- FamB %>%
    filter(LG == 1) %>%
    select(LOCUS, POSITION)

Maps <- list(LG1_A, LG1_B)

# Start writing to an output file
sink("data/LINKMAP/LG1.merge")

# merge
LG1 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG1.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "CAT", "RMSE"), sep = " ") %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG1_A <- FamA %>%
    filter(LG == 1) %>%
    select(LOCUS, POSITION)

LG1_B <- FamB %>%
    filter(LG == 1) %>%
    select(LOCUS, POSITION)

Maps <- list(LG1_A, LG1_B)

LG1 <- LPmerge(Maps, max.interval = 71, weights = NULL)

LG1[[1]] <- LG1[[1]] %>%
    mutate(LG = 1)

# choose map
write.table(LG1[[1]], "results/LG1.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG2:

```{r}

LG2_A <- FamA %>%
    filter(LG == 2) %>%
    select(LOCUS, POSITION)

LG2_B <- FamB %>%
    filter(LG == 15) %>%
    select(LOCUS, POSITION)

Maps <- list(LG2_A, LG2_B)

# Start writing to an output file
sink("data/LINKMAP/LG2.merge")

# merge
LG2 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG2.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "CAT", "RMSE"), sep = " ") %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet

```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG2_A <- FamA %>%
    filter(LG == 2) %>%
    select(LOCUS, POSITION)

LG2_B <- FamB %>%
    filter(LG == 15) %>%
    select(LOCUS, POSITION)

Maps <- list(LG2_A, LG2_B)

LG2 <- LPmerge(Maps, max.interval = 6, weights = NULL)

LG2[[1]] <- LG2[[1]] %>%
    mutate(LG = 2)

# choose map
write.table(LG2[[1]], "results/LG2.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG3:

```{r}

LG3_A <- FamA %>%
    filter(LG == 3) %>%
    select(LOCUS, POSITION)

LG3_B <- FamB %>%
    filter(LG == 8) %>%
    select(LOCUS, POSITION)

Maps <- list(LG3_A, LG3_B)

# Start writing to an output file
sink("data/LINKMAP/LG3.merge")

# merge
LG3 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG3.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "CAT", "RMSE"), sep = " ") %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

LG3_A <- FamA %>%
    filter(LG == 3) %>%
    select(LOCUS, POSITION)

LG3_B <- FamB %>%
    filter(LG == 8) %>%
    select(LOCUS, POSITION)

Maps <- list(LG3_A, LG3_B)

LG3 <- LPmerge(Maps, max.interval = 3, weights = NULL)

LG3[[1]] <- LG3[[1]] %>%
    mutate(LG = 3)

# choose map
write.table(LG3[[1]], "results/LG3.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG4:

```{r}

LG4_A <- FamA %>%
    filter(LG == 4) %>%
    select(LOCUS, POSITION)

LG4_B <- FamB %>%
    filter(LG == 2) %>%
    select(LOCUS, POSITION)

Maps <- list(LG4_A, LG4_B)

# Start writing to an output file
sink("data/LINKMAP/LG4.merge")

# merge
LG4 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG4.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "CAT", "RMSE"), sep = " ") %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG4_A <- FamA %>%
    filter(LG == 4) %>%
    select(LOCUS, POSITION)

LG4_B <- FamB %>%
    filter(LG == 2) %>%
    select(LOCUS, POSITION)

Maps <- list(LG4_A, LG4_B)

LG4 <- LPmerge(Maps, max.interval = 4, weights = NULL)

LG4[[1]] <- LG4[[1]] %>%
    mutate(LG = 4)

# choose map
write.table(LG4[[1]], "results/LG4.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG5:

```{r}

LG5_A <- FamA %>%
    filter(LG == 5) %>%
    select(LOCUS, POSITION)

LG5_B <- FamB %>%
    filter(LG == 4) %>%
    select(LOCUS, POSITION)

Maps <- list(LG5_A, LG5_B)

# Start writing to an output file
sink("data/LINKMAP/LG5.merge")

# merge
LG5 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG5.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "CAT", "RMSE"), sep = " ") %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG5_A <- FamA %>%
    filter(LG == 5) %>%
    select(LOCUS, POSITION)

LG5_B <- FamB %>%
    filter(LG == 4) %>%
    select(LOCUS, POSITION)

Maps <- list(LG5_A, LG5_B)

LG5 <- LPmerge(Maps, max.interval = 19, weights = NULL)

LG5[[1]] <- LG5[[1]] %>%
    mutate(LG = 5)

# choose map
write.table(LG5[[1]], "results/LG5.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG6:

```{r}

LG6_A <- FamA %>%
    filter(LG == 6) %>%
    select(LOCUS, POSITION)

LG6_B <- FamB %>%
    filter(LG == 3) %>%
    select(LOCUS, POSITION)

Maps <- list(LG6_A, LG6_B)

# Start writing to an output file
sink("data/LINKMAP/LG6.merge")

# merge
LG6 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG6.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "CAT", "RMSE"), sep = " ") %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG6_A <- FamA %>%
    filter(LG == 6) %>%
    select(LOCUS, POSITION)

LG6_B <- FamB %>%
    filter(LG == 3) %>%
    select(LOCUS, POSITION)

Maps <- list(LG6_A, LG6_B)

LG6 <- LPmerge(Maps, max.interval = 61, weights = NULL)

LG6[[1]] <- LG6[[1]] %>%
    mutate(LG = 6)

# choose map
write.table(LG6[[1]], "results/LG6.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG7:

```{r}

LG7_A <- FamA %>%
    filter(LG == 7) %>%
    select(LOCUS, POSITION)

LG7_B <- FamB %>%
    filter(LG == 10) %>%
    select(LOCUS, POSITION)

Maps <- list(LG7_A, LG7_B)

# Start writing to an output file
sink("data/LINKMAP/LG7.merge")

# merge
LG7 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG7.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "CAT", "RMSE"), sep = " ") %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG7_A <- FamA %>%
    filter(LG == 7) %>%
    select(LOCUS, POSITION)

LG7_B <- FamB %>%
    filter(LG == 10) %>%
    select(LOCUS, POSITION)

Maps <- list(LG7_A, LG7_B)

LG7 <- LPmerge(Maps, max.interval = 1, weights = NULL)

LG7[[1]] <- LG7[[1]] %>%
    mutate(LG = 7)

# choose map
write.table(LG7[[1]], "results/LG7.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG8:

```{r}

LG8_A <- FamA %>%
    filter(LG == 8) %>%
    select(LOCUS, POSITION)

LG8_B <- FamB %>%
    filter(LG == 12) %>%
    select(LOCUS, POSITION)

Maps <- list(LG8_A, LG8_B)

# Start writing to an output file
sink("data/LINKMAP/LG8.merge")

# merge
LG8 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG8.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "CAT", "RMSE"), sep = " ") %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG8_A <- FamA %>%
    filter(LG == 8) %>%
    select(LOCUS, POSITION)

LG8_B <- FamB %>%
    filter(LG == 12) %>%
    select(LOCUS, POSITION)

Maps <- list(LG8_A, LG8_B)

LG8 <- LPmerge(Maps, max.interval = 38, weights = NULL)

LG8[[1]] <- LG8[[1]] %>%
    mutate(LG = 8)

# choose map
write.table(LG8[[1]], "results/LG8.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG9:

```{r}

LG9_A <- FamA %>%
    filter(LG == 9) %>%
    select(LOCUS, POSITION)

LG9_B <- FamB %>%
    filter(LG == 6) %>%
    select(LOCUS, POSITION)

Maps <- list(LG9_A, LG9_B)

# Start writing to an output file
sink("data/LINKMAP/LG9.merge")

# merge
LG9 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG9.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "CAT", "RMSE"), sep = " ") %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG9_A <- FamA %>%
    filter(LG == 9) %>%
    select(LOCUS, POSITION)

LG9_B <- FamB %>%
    filter(LG == 6) %>%
    select(LOCUS, POSITION)

Maps <- list(LG9_A, LG9_B)

LG9 <- LPmerge(Maps, max.interval = 2, weights = NULL)

LG9[[1]] <- LG9[[1]] %>%
    mutate(LG = 9)

# choose map
write.table(LG9[[1]], "results/LG9.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG10:

```{r}

LG10_A <- FamA %>%
    filter(LG == 10) %>%
    select(LOCUS, POSITION)

LG10_B <- FamB %>%
    filter(LG == 21) %>%
    select(LOCUS, POSITION)

Maps <- list(LG10_A, LG10_B)

# Start writing to an output file
sink("data/LINKMAP/LG10.merge")

# merge
LG10 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG10.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG10_A <- FamA %>%
    filter(LG == 10) %>%
    select(LOCUS, POSITION)

LG10_B <- FamB %>%
    filter(LG == 21) %>%
    select(LOCUS, POSITION)

Maps <- list(LG10_A, LG10_B)

LG10 <- LPmerge(Maps, max.interval = 2, weights = NULL)

LG10[[1]] <- LG10[[1]] %>%
    mutate(LG = 10)

# choose map
write.table(LG10[[1]], "results/LG10.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG11:

```{r}

LG11_A <- FamA %>%
    filter(LG == 11) %>%
    select(LOCUS, POSITION)

LG11_B <- FamB %>%
    filter(LG == 24) %>%
    select(LOCUS, POSITION)

Maps <- list(LG11_A, LG11_B)

# Start writing to an output file
sink("data/LINKMAP/LG11.merge")

# merge
LG11 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG11.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG11_A <- FamA %>%
    filter(LG == 11) %>%
    select(LOCUS, POSITION)

LG11_B <- FamB %>%
    filter(LG == 24) %>%
    select(LOCUS, POSITION)

Maps <- list(LG11_A, LG11_B)

LG11 <- LPmerge(Maps, max.interval = 4, weights = NULL)

LG11[[1]] <- LG11[[1]] %>%
    mutate(LG = 11)

# choose map
write.table(LG11[[1]], "results/LG11.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG12:

```{r}

LG12_A <- FamA %>%
    filter(LG == 12) %>%
    select(LOCUS, POSITION)

LG12_B <- FamB %>%
    filter(LG == 5) %>%
    select(LOCUS, POSITION)

Maps <- list(LG12_A, LG12_B)

# Start writing to an output file
sink("data/LINKMAP/LG12.merge")

# merge
LG12 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG12.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "CAT", "RMSE"), sep = " ") %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG12_A <- FamA %>%
    filter(LG == 12) %>%
    select(LOCUS, POSITION)

LG12_B <- FamB %>%
    filter(LG == 5) %>%
    select(LOCUS, POSITION)

Maps <- list(LG12_A, LG12_B)

LG12 <- LPmerge(Maps, max.interval = 5, weights = NULL)

LG12[[1]] <- LG12[[1]] %>%
    mutate(LG = 12)

# choose map
write.table(LG12[[1]], "results/LG12.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG13:

```{r}

LG13_A <- FamA %>%
    filter(LG == 13) %>%
    select(LOCUS, POSITION)

LG13_B <- FamB %>%
    filter(LG == 11) %>%
    select(LOCUS, POSITION)

Maps <- list(LG13_A, LG13_B)

# Start writing to an output file
sink("data/LINKMAP/LG13.merge")

# merge
LG13 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG13.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG13_A <- FamA %>%
    filter(LG == 13) %>%
    select(LOCUS, POSITION)

LG13_B <- FamB %>%
    filter(LG == 11) %>%
    select(LOCUS, POSITION)

Maps <- list(LG13_A, LG13_B)

LG13 <- LPmerge(Maps, max.interval = 13, weights = NULL)

LG13[[1]] <- LG13[[1]] %>%
    mutate(LG = 13)

# choose map
write.table(LG13[[1]], "results/LG13.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG14:

```{r}

LG14_A <- FamA %>%
    filter(LG == 14) %>%
    select(LOCUS, POSITION)

LG14_B <- FamB %>%
    filter(LG == 22) %>%
    select(LOCUS, POSITION)

Maps <- list(LG14_A, LG14_B)

# Start writing to an output file
sink("data/LINKMAP/LG14.merge")

# merge
LG14 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG14.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG14_A <- FamA %>%
    filter(LG == 14) %>%
    select(LOCUS, POSITION)

LG14_B <- FamB %>%
    filter(LG == 22) %>%
    select(LOCUS, POSITION)

Maps <- list(LG14_A, LG14_B)

LG14 <- LPmerge(Maps, max.interval = 6, weights = NULL)

LG14[[1]] <- LG14[[1]] %>%
    mutate(LG = 14)

# choose map
write.table(LG14[[1]], "results/LG14.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG15:

```{r}

LG15_A <- FamA %>%
    filter(LG == 15) %>%
    select(LOCUS, POSITION)

LG15_B <- FamB %>%
    filter(LG == 16) %>%
    select(LOCUS, POSITION)

Maps <- list(LG15_A, LG15_B)

# Start writing to an output file
sink("data/LINKMAP/LG15.merge")

# merge
LG15 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG15.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG15_A <- FamA %>%
    filter(LG == 15) %>%
    select(LOCUS, POSITION)

LG15_B <- FamB %>%
    filter(LG == 16) %>%
    select(LOCUS, POSITION)

Maps <- list(LG15_A, LG15_B)

LG15 <- LPmerge(Maps, max.interval = 41, weights = NULL)

LG15[[1]] <- LG15[[1]] %>%
    mutate(LG = 15)

# choose map
write.table(LG15[[1]], "results/LG15.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG16:

```{r}

LG16_A <- FamA %>%
    filter(LG == 16) %>%
    select(LOCUS, POSITION)

LG16_B <- FamB %>%
    filter(LG == 14) %>%
    select(LOCUS, POSITION)

Maps <- list(LG16_A, LG16_B)

# Start writing to an output file
sink("data/LINKMAP/LG16.merge")

# merge
LG16 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG16.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG16_A <- FamA %>%
    filter(LG == 16) %>%
    select(LOCUS, POSITION)

LG16_B <- FamB %>%
    filter(LG == 14) %>%
    select(LOCUS, POSITION)

Maps <- list(LG16_A, LG16_B)

LG16 <- LPmerge(Maps, max.interval = 50, weights = NULL)

LG16[[1]] <- LG16[[1]] %>%
    mutate(LG = 16)

# choose map
write.table(LG16[[1]], "results/LG16.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG17:

```{r}

LG17_A <- FamA %>%
    filter(LG == 17) %>%
    select(LOCUS, POSITION)

LG17_B <- FamB %>%
    filter(LG == 18) %>%
    select(LOCUS, POSITION)

Maps <- list(LG17_A, LG17_B)

# Start writing to an output file
sink("data/LINKMAP/LG17.merge")

# merge
LG17 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG17.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG17_A <- FamA %>%
    filter(LG == 17) %>%
    select(LOCUS, POSITION)

LG17_B <- FamB %>%
    filter(LG == 18) %>%
    select(LOCUS, POSITION)

Maps <- list(LG17_A, LG17_B)

LG17 <- LPmerge(Maps, max.interval = 9, weights = NULL)

LG17[[1]] <- LG17[[1]] %>%
    mutate(LG = 17)

# choose map
write.table(LG17[[1]], "results/LG17.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG18:

```{r}

LG18_A <- FamA %>%
    filter(LG == 18) %>%
    select(LOCUS, POSITION)

LG18_B <- FamB %>%
    filter(LG == 17) %>%
    select(LOCUS, POSITION)

Maps <- list(LG18_A, LG18_B)

# Start writing to an output file
sink("data/LINKMAP/LG18.merge")

# merge
LG18 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG18.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG18_A <- FamA %>%
    filter(LG == 18) %>%
    select(LOCUS, POSITION)

LG18_B <- FamB %>%
    filter(LG == 17) %>%
    select(LOCUS, POSITION)

Maps <- list(LG18_A, LG18_B)

LG18 <- LPmerge(Maps, max.interval = 47, weights = NULL)

LG18[[1]] <- LG18[[1]] %>%
    mutate(LG = 18)

# choose map
write.table(LG18[[1]], "results/LG18.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG19:

```{r}

LG19_A <- FamA %>%
    filter(LG == 19) %>%
    select(LOCUS, POSITION)

LG19_B <- FamB %>%
    filter(LG == 13) %>%
    select(LOCUS, POSITION)

Maps <- list(LG19_A, LG19_B)

# Start writing to an output file
sink("data/LINKMAP/LG19.merge")

# merge
LG19 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG19.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG19_A <- FamA %>%
    filter(LG == 19) %>%
    select(LOCUS, POSITION)

LG19_B <- FamB %>%
    filter(LG == 13) %>%
    select(LOCUS, POSITION)

Maps <- list(LG19_A, LG19_B)

LG19 <- LPmerge(Maps, max.interval = 3, weights = NULL)

LG19[[1]] <- LG19[[1]] %>%
    mutate(LG = 19)

# choose map
write.table(LG19[[1]], "results/LG19.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG20:

```{r}

LG20_A <- FamA %>%
    filter(LG == 20) %>%
    select(LOCUS, POSITION)

LG20_B <- FamB %>%
    filter(LG == 7) %>%
    select(LOCUS, POSITION)

Maps <- list(LG20_A, LG20_B)

# Start writing to an output file
sink("data/LINKMAP/LG20.merge")

# merge
LG20 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG20.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

LG20_A <- FamA %>%
    filter(LG == 20) %>%
    select(LOCUS, POSITION)

LG20_B <- FamB %>%
    filter(LG == 7) %>%
    select(LOCUS, POSITION)

Maps <- list(LG20_A, LG20_B)

LG20 <- LPmerge(Maps, max.interval = 1, weights = NULL)

LG20[[1]] <- LG20[[1]] %>%
    mutate(LG = 20)

# choose map
write.table(LG20[[1]], "results/LG20.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG21:

```{r}

LG21_A <- FamA %>%
    filter(LG == 21) %>%
    select(LOCUS, POSITION)

LG21_B <- FamB %>%
    filter(LG == 19) %>%
    select(LOCUS, POSITION)

Maps <- list(LG21_A, LG21_B)

# Start writing to an output file
sink("data/LINKMAP/LG21.merge")

# merge
LG21 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG21.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG21_A <- FamA %>%
    filter(LG == 21) %>%
    select(LOCUS, POSITION)

LG21_B <- FamB %>%
    filter(LG == 19) %>%
    select(LOCUS, POSITION)

Maps <- list(LG21_A, LG21_B)

LG21 <- LPmerge(Maps, max.interval = 2, weights = NULL)

LG21[[1]] <- LG21[[1]] %>%
    mutate(LG = 21)

# choose map
write.table(LG21[[1]], "results/LG21.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG22:

```{r}

LG22_A <- FamA %>%
    filter(LG == 22) %>%
    select(LOCUS, POSITION)

LG22_B <- FamB %>%
    filter(LG == 9) %>%
    select(LOCUS, POSITION)

Maps <- list(LG22_A, LG22_B)

# Start writing to an output file
sink("data/LINKMAP/LG22.merge")

# merge
LG22 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG22.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG22_A <- FamA %>%
    filter(LG == 22) %>%
    select(LOCUS, POSITION)

LG22_B <- FamB %>%
    filter(LG == 9) %>%
    select(LOCUS, POSITION)

Maps <- list(LG22_A, LG22_B)

LG22 <- LPmerge(Maps, max.interval = 1, weights = NULL)

LG22[[1]] <- LG22[[1]] %>%
    mutate(LG = 22)

# choose map
write.table(LG22[[1]], "results/LG22.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG23:

```{r}

LG23_A <- FamA %>%
    filter(LG == 23) %>%
    select(LOCUS, POSITION)

LG23_B <- FamB %>%
    filter(LG == 20) %>%
    select(LOCUS, POSITION)

Maps <- list(LG23_A, LG23_B)

# Start writing to an output file
sink("data/LINKMAP/LG23.merge")

# merge
LG23 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG23.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG23_A <- FamA %>%
    filter(LG == 23) %>%
    select(LOCUS, POSITION)

LG23_B <- FamB %>%
    filter(LG == 20) %>%
    select(LOCUS, POSITION)

Maps <- list(LG23_A, LG23_B)

LG23 <- LPmerge(Maps, max.interval = 20, weights = NULL)

LG23[[1]] <- LG23[[1]] %>%
    mutate(LG = 23)

# choose map
write.table(LG23[[1]], "results/LG23.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Merge maps for LG24:

```{r}

LG24_A <- FamA %>%
    filter(LG == 24) %>%
    select(LOCUS, POSITION)

LG24_B <- FamB %>%
    filter(LG == 23) %>%
    select(LOCUS, POSITION)

Maps <- list(LG24_A, LG24_B)

# Start writing to an output file
sink("data/LINKMAP/LG24.merge")

# merge
LG24 <- LPmerge(Maps, max.interval = 1:100, weights = NULL)

# Stop writing to the file
sink()

```

Compare at stats to choose K by minimizing RMSE for each map (mean RMSE, standard deviation):

```{r}

# load file
merge_stats <- readLines("data/LINKMAP/LG24.merge")

# markers with constraints
map_extract <- as.data.frame(grep("Map", merge_stats, value = TRUE)) %>%
  rename(temp = `grep("Map", merge_stats, value = TRUE)`) %>%
  separate(temp, into = c("x", "y"), sep = "]") %>%
  separate(y, into = c("MAP", "LOCUS"), sep = ":") %>%
    select(MAP, LOCUS) %>%
    mutate(MAP = str_replace(MAP, "\"", " ")) %>%
    mutate(LOCUS = str_replace(LOCUS, "\"", " ")) %>%
    arrange(MAP)

# extract all RMSE values
map_rmse <- as.data.frame(grep("]" , merge_stats, invert = TRUE, value = TRUE)) %>%
    rename(temp = `grep("]", merge_stats, invert = TRUE, value = TRUE)`) %>%
    mutate(temp = str_trim(temp)) %>%
    droplevels()

# RMSE values for mean
rmse_mean <- map_rmse %>%
    filter(grepl("mean", temp)) %>%
    separate(temp, into = c("x", "RMSE"), sep = 2) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = "mean") %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values for SD
rmse_sd <- map_rmse %>%
    filter(grepl("sd", temp)) %>%
    separate(temp, into = c("x", "y"), sep = 4) %>%
    separate(y, into = c("CAT", "RMSE"), sep = 3) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE) %>%
    mutate(K = c(1:100))

# RMSE values map 1 & 2
rmse_map <- map_rmse %>%
    filter(!grepl("sd | mean", temp)) %>%
    filter(!grepl("map", temp)) %>%
    separate(temp, into = c("CAT", "y"), sep = c(1)) %>%
    separate(y, into = c("y", "RMSE"), sep = -6) %>%
    mutate(RMSE = parse_number(RMSE)) %>%
    mutate(CAT = str_trim(CAT)) %>%
    select(CAT, RMSE)

rmse_map1 <- rmse_map %>%
    filter(CAT == 1) %>%
    mutate(K = c(1:100))

rmse_map2 <- rmse_map %>%
    filter(CAT == 2) %>%
    mutate(K = c(1:100))

# create data frame with all rmse values
merge <- bind_rows(rmse_mean, rmse_sd)
merge <- bind_rows(merge, rmse_map1)
merge <- bind_rows(merge, rmse_map2)

ggplot(merge, aes(x = K, y = RMSE)) +
    geom_point() +
    facet_grid(CAT ~ ., scales = "free") +
    theme_facet


```

Write chosen consensus map to file:

```{r}

# merge map for chosen K value
LG24_A <- FamA %>%
    filter(LG == 24) %>%
    select(LOCUS, POSITION)

LG24_B <- FamB %>%
    filter(LG == 23) %>%
    select(LOCUS, POSITION)

Maps <- list(LG24_A, LG24_B)

LG24 <- LPmerge(Maps, max.interval = 3, weights = NULL)

LG24[[1]] <- LG24[[1]] %>%
    mutate(LG = 24)

# choose map
write.table(LG24[[1]], "results/LG24.consensus", quote = FALSE, row.names = FALSE, col.names = c("LOCUS", "POS", "POS_FamA", "POS_FamB", "LG"))

```

## Finalize southern flounder linkage map

Import all merged consensus linkage groups and combine into one map file.

```{r}

filenames <- list.files(path = "results", pattern = "*.consensus")

# import data
for (i in filenames){
  filepath <- file.path("results/", paste(i))
  assign(i, read.table(filepath, sep = "", header = TRUE))
  }


dflist_consensus <- lapply(ls(pattern = "*.consensus"), get)

for (df in dflist_consensus)

map <- ldply(dflist_consensus, data.frame) %>%
    select(LOCUS, LG, POS)

write.table(map, "results/SFL.map", 
            col.names = TRUE, row.names = FALSE, sep = "\t", quote = FALSE)

```


# Compare stats for sex-specific, family and consensus linkage maps

Load sex-specific maps:

```{r}

FamA_F.map <- read.delim("results/FamA_F.map", 
                   stringsAsFactors = FALSE, sep = "",
                   col.names = c("LOCUS", "LG_FamA_F", "POS"))

FamA_M.map <- read.delim("results/FamA_M.map", 
                   stringsAsFactors = FALSE, sep = "",
                   col.names = c("LOCUS", "LG_FamA_M", "POS"))

FamB_F.map <- read.delim("results/FamB_F.map", 
                   stringsAsFactors = FALSE, sep = "",
                   col.names = c("LOCUS", "LG_FamB_F", "POS"))

FamB_M.map <- read.delim("results/FamB_M.map", 
                   stringsAsFactors = FALSE, sep = "",
                   col.names = c("LOCUS", "LG_FamB_M", "POS"))

```

Load family maps:

```{r}

FamA.map <- read.delim("results/FamA_family.map", 
                   stringsAsFactors = FALSE, sep = "",
                   col.names = c("LOCUS", "LG_FamA", "POS")) %>%
    mutate(LOCUS = str_sub(LOCUS, 2))

FamB.map <- read.delim("results/FamB_family.map", 
                   header = TRUE, stringsAsFactors = FALSE, sep = "", 
                   col.names = c("LOCUS", "LG_FamB", "POS")) %>%
    mutate(LOCUS = str_sub(LOCUS, 2))

```

Load consensus map:

```{r}

SFL.map <- read.delim("results/SFL.map", 
                   header = TRUE, stringsAsFactors = FALSE, sep = "", 
                   col.names = c("LOCUS", "LG", "POS")) %>%
    mutate(LOCUS = str_sub(LOCUS, 2))

```

## Identify corresponding LG:


```{r}

LGs <- inner_join(FamA_F.map, FamA_M.map, by = "LOCUS")

LGs <- inner_join(FamA.map, LGs, by = "LOCUS")

LGs <- inner_join(FamB_F.map, LGs, by = "LOCUS")

LGs <- inner_join(FamB_M.map, LGs, by = "LOCUS")

LGs <- inner_join(FamB.map, LGs, by = "LOCUS")

LGs <- inner_join(SFL.map, LGs, by = "LOCUS") %>%
    select(LG, LG_FamB, LG_FamB_F, LG_FamB_M, LG_FamA, LG_FamA_F, LG_FamA_M) %>%
    distinct(LG, .keep_all = TRUE) %>%
    add_row(LG = 12, 
            LG_FamB = 5,
            LG_FamB_F = 5,
            LG_FamB_M = 6,
            LG_FamA = 12,
            LG_FamA_F = 16,
            LG_FamA_M = 2)

write.table(LGs, "results/LGs", row.names = FALSE, col.names = TRUE, quote = FALSE)

```

## Compare length and number of loci:

Number of markers per linkage group:

```{r}

temp <- count(SFL.map, LG)

temp <- count(FamA.map, LG_FamA)

temp <- count(FamA_F.map, LG_FamA_F)

temp <- count(FamA_M.map, LG_FamA_M)

temp <- count(FamB.map, LG_FamB)

temp <- count(FamB_F.map, LG_FamB_F)

temp <- count(FamB_M.map, LG_FamB_M)

```

Length per linkage group:

```{r}

temp <- SFL.map %>%
    group_by(LG) %>%
    summarize(length = max(POS))

temp <- FamA.map %>%
    group_by(LG_FamA) %>%
    summarize(length = max(POS))

temp <- FamA_F.map %>%
    group_by(LG_FamA_F) %>%
    summarize(length = max(POS))

temp <- FamA_M.map %>%
    group_by(LG_FamA_M) %>%
    summarize(length = max(POS))

temp <- FamB.map %>%
    group_by(LG_FamB) %>%
    summarize(length = max(POS))

temp <- FamB_F.map %>%
    group_by(LG_FamB_F) %>%
    summarize(length = max(POS))

temp <- FamB_M.map %>%
    group_by(LG_FamB_M) %>%
    summarize(length = max(POS))

```

## Plot linear linkage maps

Plot linkage map:

```{r}

# plot with ggplot
ggplot(SFL.map, aes(x = LG, y = POS)) +
  geom_path(aes(group = LG), size = 9, lineend = "round", col = "black") +
  geom_path(aes(group = LG), size = 7.5, lineend = "round", col = "grey95") +
  geom_rect(aes(xmin = LG-0.35, xmax = LG+0.35, 
                ymin = POS-0.025, ymax = POS+0.025),
            color = "black", fill = "black") +
  scale_x_continuous(breaks = c(1:24)) +
  scale_y_continuous(breaks = c(0, 10, 20, 30, 40, 50, 60, 70, 80)) +
  # scale_fill_manual(name="Marker Type",
  #                   breaks=c("est", "msat", "snp"),
  #                   labels=c("EST-SSR", "Microsatellite", "RAD Locus"),
  #                   values=c('#9e0142', '#a1d99b', '#2b8cbe') +
  labs(x = "\nLinkage Group", y = "Position (cM)\n") +
  theme_classic() +
  theme(axis.title = element_text(size = 16),
        axis.title.y = element_text(vjust = 1.5),
        axis.line.x = element_blank(),
        axis.line.y = element_line(color = "black", size = 1),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_line(color = "black", size = 1),
        axis.ticks.length = unit(0.2, "cm"),
        axis.text.x = element_text(size = 14),
        axis.text.y = element_text(size = 12),
        
        legend.position = "bottom",
        
        panel.background = element_rect(fill = "white", color = NA), 
        panel.border = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        strip.background = element_rect(fill = "grey95", color = "black"), 
        strip.text.x = element_text(size = 16),
        strip.text.y = element_text(size = 16))

```

Linkage map Family A:

```{r}

FamA.map <- read.delim("data/LINKMAP/FamA.fullmap", 
                   stringsAsFactors = FALSE, sep = "", 
                   col.names = c("LG", "LOCUS", "POSITION"))

# plot with ggplot
ggplot(FamA.map, aes(x = LG, y = POSITION)) +
  geom_path(aes(group = LG), size = 9, lineend = "round", col = "black") +
  geom_path(aes(group = LG), size = 7.5, lineend = "round", col = "grey95") +
  geom_rect(aes(xmin = LG-0.35, xmax = LG+0.35, 
                ymin = POSITION-0.025, ymax = POSITION+0.025),
            color = "black", fill = "black") +
  scale_x_continuous(breaks = c(1:24)) +
  scale_y_continuous(breaks = c(0, 10, 20, 30, 40, 50, 60, 70, 80)) +
  # scale_fill_manual(name="Marker Type",
  #                   breaks=c("est", "msat", "snp"),
  #                   labels=c("EST-SSR", "Microsatellite", "RAD Locus"),
  #                   values=c('#9e0142', '#a1d99b', '#2b8cbe') +
  labs(x = "\nLinkage Group", y = "Position (cM)\n") +
  theme_classic() +
  theme(axis.title = element_text(size = 16),
        axis.title.y = element_text(vjust = 1.5),
        axis.line.x = element_blank(),
        axis.line.y = element_line(color = "black", size = 1),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_line(color = "black", size = 1),
        axis.ticks.length = unit(0.2, "cm"),
        axis.text.x = element_text(size = 14),
        axis.text.y = element_text(size = 12),
        
        legend.position = "bottom",
        
        panel.background = element_rect(fill = "white", color = NA), 
        panel.border = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        strip.background = element_rect(fill = "grey95", color = "black"), 
        strip.text.x = element_text(size = 16),
        strip.text.y = element_text(size = 16))

```

Linkage Map Family B:

```{r}


FamB.map <- read.delim("data/LINKMAP/FamB.fullmap", 
                   header = TRUE, stringsAsFactors = FALSE, sep = "", 
                   col.names = c("LG", "LOCUS", "POSITION"))

ggplot(FamB.map, aes(x = LG, y = POSITION)) +
  geom_path(aes(group = LG), size = 9, lineend = "round", col = "black") +
  geom_path(aes(group = LG), size = 7.5, lineend = "round", col = "grey95") +
  geom_rect(aes(xmin = LG-0.35, xmax = LG+0.35, 
                ymin = POSITION-0.025, ymax = POSITION+0.025),
            color = "black", fill = "black") +
  scale_x_continuous(breaks = c(1:24)) +
  scale_y_continuous(breaks = c(0, 10, 20, 30, 40, 50, 60, 70, 80)) +
  # scale_fill_manual(name="Marker Type",
  #                   breaks=c("est", "msat", "snp"),
  #                   labels=c("EST-SSR", "Microsatellite", "RAD Locus"),
  #                   values=c('#9e0142', '#a1d99b', '#2b8cbe') +
  labs(x = "\nLinkage Group", y = "Position (cM)\n") +
  theme_classic() +
  theme(axis.title = element_text(size = 16),
        axis.title.y = element_text(vjust = 1.5),
        axis.line.x = element_blank(),
        axis.line.y = element_line(color = "black", size = 1),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_line(color = "black", size = 1),
        axis.ticks.length = unit(0.2, "cm"),
        axis.text.x = element_text(size = 14),
        axis.text.y = element_text(size = 12),
        
        legend.position = "bottom",
        
        panel.background = element_rect(fill = "white", color = NA), 
        panel.border = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        strip.background = element_rect(fill = "grey95", color = "black"), 
        strip.text.x = element_text(size = 16),
        strip.text.y = element_text(size = 16))

```

## Plot circular linkage maps:

### Sex-specific maps

FamA female

```{r}

# determine length of each LG
by_lg <- group_by(FamA_F.map, LG_FamA_F) %>%
  top_n(1, POS)

# format karyotype file
kary <- ungroup(by_lg) %>%
    select(LG_FamA_F, POS) %>%
    unique() %>% # not sure why some LGs are duplicates?
    mutate(COLOR = "black") %>%
    mutate(id = paste("lg", LG_FamA_F, sep=''), 
           label = paste("LG", LG_FamA_F, sep=''),
           chr = "chr", 
           dash = "-", 
           start = "0", 
           end = as.integer(POS*500000), 
           color = "black") %>%
    select(chr, dash, id, label, start, end, COLOR)

# output karyotype data file
write.table(kary, file = "data/LINKMAP/FamA-F.karyotype", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

markers <- FamA_F.map %>%
    mutate(chr = paste("lg", LG_FamA_F, sep=''),
           start = as.integer(POS*500000), 
           end = as.integer(POS*500000)) %>%
    select(chr, start, end)

write.table(markers, file = "data/LINKMAP/FamA-F.loci", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

system('perl "scr/circos-0.66/bin/circos" -conf data/LINKMAP/circos-A-f.conf -outputdir fig -outputfile FamA-f')

```

FamA male

```{r}

# determine length of each LG
by_lg <- group_by(FamA_M.map, LG_FamA_M) %>%
  top_n(1, POS)

# format karyotype file
kary <- ungroup(by_lg) %>%
    select(LG_FamA_M, POS) %>%
    unique() %>% # not sure why some LGs are duplicates?
    mutate(COLOR = "black") %>%
    mutate(id = paste("lg", LG_FamA_M, sep=''), 
           label = paste("LG", LG_FamA_M, sep=''),
           chr = "chr", 
           dash = "-", 
           start = "0", 
           end = as.integer(POS*500000), 
           color = "black") %>%
    select(chr, dash, id, label, start, end, COLOR)

# output karyotype data file
write.table(kary, file = "data/LINKMAP/FamA-M.karyotype", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

markers <- FamA_M.map %>%
    mutate(chr = paste("lg", LG_FamA_M, sep=''),
           start = as.integer(POS*500000), 
           end = as.integer(POS*500000)) %>%
    select(chr, start, end)

write.table(markers, file = "data/LINKMAP/FamA-M.loci", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

system('perl "scr/circos-0.66/bin/circos" -conf data/LINKMAP/circos-A-m.conf -outputdir fig -outputfile FamA-m')

```

FamB female

```{r}

# determine length of each LG
by_lg <- group_by(FamB_F.map, LG_FamB_F) %>%
  top_n(1, POS)

# format karyotype file
kary <- ungroup(by_lg) %>%
    select(LG_FamB_F, POS) %>%
    unique() %>% # not sure why some LGs are duplicates?
    mutate(COLOR = "black") %>%
    mutate(id = paste("lg", LG_FamB_F, sep=''), 
           label = paste("LG", LG_FamB_F, sep=''),
           chr = "chr", 
           dash = "-", 
           start = "0", 
           end = as.integer(POS*500000), 
           color = "black") %>%
    select(chr, dash, id, label, start, end, COLOR)

# output karyotype data file
write.table(kary, file = "data/LINKMAP/FamB-F.karyotype", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

markers <- FamB_F.map %>%
    mutate(chr = paste("lg", LG_FamB_F, sep=''),
           start = as.integer(POS*500000), 
           end = as.integer(POS*500000)) %>%
    select(chr, start, end)

write.table(markers, file = "data/LINKMAP/FamB-F.loci", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

system('perl "scr/circos-0.66/bin/circos" -conf data/LINKMAP/circos-B-f.conf -outputdir fig -outputfile FamB-f')

```

FamB male

```{r}

# determine length of each LG
by_lg <- group_by(FamB_M.map, LG_FamB_M) %>%
  top_n(1, POS)

# format karyotype file
kary <- ungroup(by_lg) %>%
    select(LG_FamB_M, POS) %>%
    unique() %>% # not sure why some LGs are duplicates?
    mutate(COLOR = "black") %>%
    mutate(id = paste("lg", LG_FamB_M, sep=''), 
           label = paste("LG", LG_FamB_M, sep=''),
           chr = "chr", 
           dash = "-", 
           start = "0", 
           end = as.integer(POS*500000), 
           color = "black") %>%
    select(chr, dash, id, label, start, end, COLOR)

# output karyotype data file
write.table(kary, file = "data/LINKMAP/FamB-M.karyotype", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

markers <- FamB_M.map %>%
    mutate(chr = paste("lg", LG_FamB_M, sep=''),
           start = as.integer(POS*500000), 
           end = as.integer(POS*500000)) %>%
    select(chr, start, end)

write.table(markers, file = "data/LINKMAP/FamB-M.loci", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

system('perl "scr/circos-0.66/bin/circos" -conf data/LINKMAP/circos-B-m.conf -outputdir fig -outputfile FamB-m')

```

### Family Maps

FamA

```{r}

# determine length of each LG
by_lg <- group_by(FamA.map, LG_FamA) %>%
  top_n(1, POS)

# format karyotype file
kary <- ungroup(by_lg) %>%
    select(LG_FamA, POS) %>%
    unique() %>% # not sure why some LGs are duplicates?
    mutate(COLOR = "black") %>%
    mutate(id = paste("lg", LG_FamA, sep=''), 
           label = paste("LG", LG_FamA, sep=''),
           chr = "chr", 
           dash = "-", 
           start = "0", 
           end = as.integer(POS*500000), 
           color = "black") %>%
    select(chr, dash, id, label, start, end, COLOR)

# output karyotype data file
write.table(kary, file = "data/LINKMAP/FamA.karyotype", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

markers <- FamA.map %>%
    mutate(chr = paste("lg", LG_FamA, sep=''),
           start = as.integer(POS*500000), 
           end = as.integer(POS*500000)) %>%
    select(chr, start, end)

write.table(markers, file = "data/LINKMAP/FamA.loci", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

system('perl "scr/circos-0.66/bin/circos" -conf data/LINKMAP/circos-A.conf -outputdir fig -outputfile FamA')

```

FamB

```{r}

# determine length of each LG
by_lg <- group_by(FamB.map, LG_FamB) %>%
  top_n(1, POS)

# format karyotype file
kary <- ungroup(by_lg) %>%
    select(LG_FamB, POS) %>%
    unique() %>% # not sure why some LGs are duplicates?
    mutate(COLOR = "black") %>%
    mutate(id = paste("lg", LG_FamB, sep=''), 
           label = paste("LG", LG_FamB, sep=''),
           chr = "chr", 
           dash = "-", 
           start = "0", 
           end = as.integer(POS*500000), 
           color = "black") %>%
    select(chr, dash, id, label, start, end, COLOR)

# output karyotype data file
write.table(kary, file = "data/LINKMAP/FamB.karyotype", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

markers <- FamB.map %>%
    mutate(chr = paste("lg", LG_FamB, sep=''),
           start = as.integer(POS*500000), 
           end = as.integer(POS*500000)) %>%
    select(chr, start, end)

write.table(markers, file = "data/LINKMAP/FamB.loci", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

system('perl "scr/circos-0.66/bin/circos" -conf data/LINKMAP/circos-B.conf -outputdir fig -outputfile FamB')

```