---
title: "Comparative Genomics Southern Flounder"
output:
  html_notebook:
    code_folding: hide
    df_print: paged
    highlight: kate
    theme: flatly
    toc: yes
  html_document:
    toc: yes
---

```{r load libraries, message=FALSE, warning=FALSE}

knitr::opts_chunk$set(warning = FALSE, message = FALSE)

source("lib/libraries.R")
source("lib/ggplot.R")

```

# Synteny mapped loci

Output from synteny mapper `all.synteny.mapped.out`.

```{r}

# load map data
map <- read.table("data/SYNTENY/map.tab", header = TRUE)

# load synteny data
data <- read.table("data/SYNTENY/all.blocks.tab", header=TRUE)

# chrom lengths fish genomes
chrom <- read.table("data/SYNTENY/chr_lengths.txt", header = TRUE, stringsAsFactors = FALSE)

# load synteny data
syn_blocks <- read.table("data/SYNTENY/all.blocks.tab", header=TRUE)

# load synteny mapped loci
syn_loci <- read.table("data/SYNTENY/all.synteny.mapped.out", header=TRUE)

```

# write Karyotype files

Create karyotype file for southern flounder consensus map.

```{r}

# determine length of each LG
by_lg <- group_by(map, LG) %>%
  top_n(1, POSITION)

# format karyotype file
kary <- ungroup(by_lg) %>%
  select(LG, POSITION) %>%
  unique() %>% # not sure why some LGs are duplicates?
  mutate(COLOR = "black") %>%
  mutate(id = paste("lg", LG, sep=''), 
         label = paste("LG", LG, sep=''),
         chr = "chr", 
         dash = "-", 
         start = "0", 
         end = as.integer(POSITION * 500000), 
         color = "grey") %>%
  select(chr, dash, id, label, start, end, COLOR)

# output karyotype data file
write.table(kary, file = "scr/CIRCOS/SFL_karyotype.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Write markers file (loci positions):

```{r}

# ----- WRITE MARKERS FILE -------------------------------------------------------------

markers <- mutate(map, chr = paste("soc", lg, sep=''), start = as.integer(pos * 500000), end = as.integer(pos * 500000))
markers_final <- markers[,c(4, 5, 6, 1)]

write.table(markers_final, file = "data/circos/markers.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Write tile track file

```{r}

# Get a list of the comparison species

species <- as.vector(unique(data$COMP_SPECIES))

for (i in 1:length(species)) {
  spec <- filter(data, COMP_SPECIES == species[i])
  spec <- mutate(spec, chr = paste("lg", MAP_LG, sep=''), start = as.integer(MAP_START * 500000), end = as.integer(MAP_STOP * 500000))
  tiles <- select(spec, chr, start, end)
  write.table(tiles, file = paste("scr/CIRCOS/data/", species[i], ".txt", sep = ''), sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
}

# Write a tile file for the synteny mapped loci

syn_loci <- mutate(syn, chr = paste("lg", LG, sep=''), start = as.integer(LEFT_POS * 500000), end = as.integer(RIGHT_POS * 500000))
loci <- select(syn_loci, chr, start, end)
write.table(loci, file = "data/circos/syn_loci.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Run circos to plot:

```{r}

system('perl "C:\Program Files\Circos\circos-0.69-2\bin\circos" -conf data\circos\circos.conf -outputdir fig -outputfile fig2')

```

# Create input files

```{r}

# --- Southern Flounder Linkage map -----------------------------------------------------

# determine length of each LG
by_lg <- group_by(map, LG) %>%
  top_n(1, POSITION)

# format karyotype file
kary <- ungroup(by_lg) %>%
  select(LG, POSITION) %>%
  unique() %>% # not sure why some LGs are duplicates?
  mutate(id = paste("lg", LG, sep=''),
         label = paste("LG", LG, sep=''),
         chr = "chr",
         dash = "-",
         start = "0",
         end = as.integer(POSITION * 500000),
         color = "grey") %>%
  select(chr, dash, id, label, start, end, color)

# output karyotype data file
write.table(kary, file = "scr/CIRCOS/SFL_karyotype.txt", 
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```



```{r}

# --- European Seabass (dlab) ------------------------------------------------------------

# format file
kary <- chrom %>%
  filter(species == "dlab") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "grey") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "scr/CIRCOS/dlab_karyotype.txt",
              sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)


```



```{r}
# --- Stickleback (gacu) -------------------------------------------------------------

# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "gacu") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "grey") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "scr/CIRCOS/gacu_karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
  

```



```{r}
# --- Asian Seabass (lcal) -------------------------------------------------------------

# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "lcal") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "grey") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "scr/CIRCOS/lcal_karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```



```{r}
# --- Nile tilapia (onil) -------------------------------------------------------------

# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "onil") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "grey") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "scr/CIRCOS/onil_karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```



```{r}
# --- spotted green puffer (tnig) -------------------------------------------------------------

# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "tnig") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "grey") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "scr/CIRCOS/tnig_karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```



```{r}
# --- Fugu (trub) -------------------------------------------------------------
  
# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "trub") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "grey") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "scr/CIRCOS/trub_karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```



```{r}
# --- Zebrafish (drer) -------------------------------------------------------------

# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "drer") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "grey") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "scr/CIRCOS/drer_karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```



```{r}
# ----- WRITE LOCUS TRACK FILES -----------------------------------------------------

# format SNP locus track file (use for highlight plot blocks)
SNPloci <- map %>%
  mutate(chr = paste("LG", LG, sep=''), 
         start = as.integer(POSITION * 500000), 
         end = as.integer(POSITION * 500000),
         color = "black") %>%
  select(chr, start, end, color)

# write input file
write.table(SNPloci, file = "scr/CIRCOS/SNPloci.txt", 
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```



```{r}
# ----- WRITE SYNTENY BLOCK TILE TRACK FILES ---------------------------------------------

tiles-a <- synblocks %>%
  filter(MAP_START < MAP_STOP) %>%
  mutate(start = MAP_START*500000,
         end = MAP_STOP*500000)

tiles-b <- synblocks %>%
  filter(MAP_START > MAP_STOP) %>%
  mutate(start = MAP_STOP*500000,
         end = MAP_START*500000)

tiles <- bind_rows()


# Get a list of the comparison species
species <- as.vector(unique(synblocks$COMP_SPECIES))

# write tile track file for synteny blocks for each species
for (i in 1:length(species)) {
  # create data frame with synteny blocks
  tiles <- synblocks %>%
    filter(COMP_SPECIES == species[i]) %>%
    mutate(chr = paste("LG", MAP_LG, sep=''), 
           start = as.integer(MAP_START * 500000), 
           end = as.integer(MAP_STOP * 500000)) %>%
    select(chr, start, end)
  # write tile track file
  write.table(tiles, file = paste("scr/CIRCOS/", species[i], "_synt_blocks.txt", sep = ''), 
              sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
}

```



```{r}
# Write a tile track files for the synteny mapped loci

# create one files with all information and then split (same as species?)

# Format files
# syn_loci <- syn_loci %>%
#   mutate(chr = paste("lg", LG, sep=''), 
#          start = as.integer(LEFT_POS * 500000),
#          end = as.integer(RIGHT_POS * 500000)) %>%
#   select(chr, start, end)

# write outputfile
# write.table(loci, file = "data/circos/syn_loci.txt", 
#             sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```



```{r}
# ----- WRITE LINK FILES ---------------------------------------------

# Format file
dlab_synblocks <- synt_blocks %>%
  filter(COMP_SPECIES == "dlab") %>%
  mutate(lg = paste("lg", MAP_LG, sep=''), 
         lgstart = as.integer(MAP_START * 500000),
         lgend = as.integer(MAP_STOP * 500000),
         chr = COMP_CHR,
         chrstart = COMP_START,
         chrend = COMP_END) %>%
  select(lg, lgstart, lgend, chr, chrstart, chrend)

write.table(dlab_synblocks, file = "scr/CIRCOS/dlab_synteny.txt", 
                         sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

# Run circos

```{r}

# ----- RUN CIRCOS ---------------------------------------------------------------------
# run from within R 
# figures end up in R working directory, png and svg format; same name as main configuration file
system('perl "C:/Program Files/Circos/circos-0.69-2/bin/circos" -conf scr/CIRCOS/circos.conf')

# create basic ideogram
system('perl "C:/Program Files/Circos/circos-0.69-2/bin/circos" -conf scr/CIRCOS/circos_ideogram_basic.conf')

# create ideogram with SNPs (highlight block)
system('perl "C:/Program Files/Circos/circos-0.69-2/bin/circos" -conf scr/CIRCOS/circos_ideogram_SNPs.conf')

# create ideogram with tile tracks for synteny blocks
system('perl "C:/Program Files/Circos/circos-0.69-2/bin/circos" -conf scr/CIRCOS/circos_tiles_syntenyblocks.conf')

# create ideogram with link tracks for synteny blocks
system('perl "C:/Program Files/Circos/circos-0.69-2/bin/circos" -conf scr/CIRCOS/circos_links_syntenyblocks.conf')

system('perl "C:/Program Files/Circos/circos-0.69-2/bin/circos" -conf scr/CIRCOS/chris.conf')

```

