---
title: "Comparative Genomics Southern Flounder"
output:
  html_notebook:
    code_folding: hide
    df_print: paged
    highlight: kate
    theme: flatly
    toc: yes
  html_document:
    toc: yes
---

```{r load libraries, message=FALSE, warning=FALSE}

knitr::opts_chunk$set(warning = FALSE, message = FALSE)

source("lib/libraries.R")
source("lib/ggplot.R")

```

# Load data

Load consensus linkage map data, chromosome lengths per species, data set of synteny blocks identified per species and loci that have been synteny mapped.

```{r}

# load map data
map <- read.table("data/SYNTENY/map.tab", 
                  header=TRUE, stringsAsFactors = FALSE, quote = "")

# chrom lengths fish genomes
chrom <- read.table("data/SYNTENY/chr_lengths.txt", 
                    header = TRUE, stringsAsFactors = FALSE)

# load synteny data
syn_blocks <- read.table("data/SYNTENY/all.blocks.tab", 
                         header=TRUE, stringsAsFactors = FALSE)

# load synteny mapped loci
syn_loci <- read.table("data/SYNTENY/all.synteny.mapped.out",
                       header=TRUE, stringsAsFactors = FALSE)

# create dataframe with LOCUS and ACCESSION No. from FASTA file
flounderloci_IDs <- read.delim("data/SYNTENY/syn-map.loci", sep = "|", header = FALSE) %>%
  rename(LOCUS = V2, ACCESSION = V4) %>%
  separate(ACCESSION, c("ACCESSION", "temp"), sep = -3) %>%
  select(LOCUS, ACCESSION)

```

# Write input files

## Karyotype files

Create karyotype file for southern flounder consensus map:

```{r}

# determine length of each LG
by_lg <- group_by(map, LG) %>%
  top_n(1, POS)

# format karyotype file
kary <- ungroup(by_lg) %>%
    select(LG, POS) %>%
    unique() %>% # not sure why some LGs are duplicates?
    mutate(COLOR = "black") %>%
    mutate(id = paste("lg", LG, sep=''), 
           label = paste("LG", LG, sep=''),
           chr = "chr", 
           dash = "-", 
           start = "0", 
           end = as.integer(POS*500000), 
           color = "black") %>%
    select(chr, dash, id, label, start, end, COLOR)

# output karyotype data file
write.table(kary, file = "data/SYNTENY/SFL.karyotype.txt", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

```

Create karyotype file for European seabass genome:

```{r}

# format file
kary <- chrom %>%
  filter(species == "dlab") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "black") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "data/SYNTENY/dlab.karyotype.txt",
              sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Create karyotype file for stickleback (gacu):

```{r}

# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "gacu") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "black") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "data/SYNTENY/gacu.karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
  
```

Create karyotype file for Asian seabass (lcal):

```{r}

# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "lcal") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "black") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "data/SYNTENY/lcal.karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Create karyotype file for Nile tilapia (onil):

```{r}

# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "onil") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "black") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "data/SYNTENY/onil.karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Create karyotype file for spotted green puffer (tnig):

```{r}

# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "tnig") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "black") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "data/SYNTENY/tnig.karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Create karyotype file for fugu (trub):

```{r}

# create data frame with chromosome lengths
kary <- chrom %>%
  filter(species == "trub") %>%
  mutate(chr = "chr",
         dash = "-",
         id = chrom,
         label = chrom,
         start = "0", 
         end = length,
         color = "black") %>%
  select(chr, dash, id, label, start, end, color)

# write karyotype track file
write.table(kary, "data/SYNTENY/trub.karyotype.txt",
            sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

## Tile track files

### SNP loci southern flounder

Create markers file with positions of mapped SNP loci on each linkage group (chromosome) in consensus linkage map for southern flounder.

```{r}

markers <- map %>%
    mutate(chr = paste("lg", LG, sep=''),
           start = as.integer(POS*500000), 
           end = as.integer(POS*500000)) %>%
    select(chr, start, end)

write.table(markers, file = "data/SYNTENY/SNPmarkers.loci", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

```

### Write tile track files for synteny blocks:

Create a file with synteny blocks identified for each species of comparison.

```{r}

# write tile track for synteny blocks for each species

species <- as.vector(unique(syn_blocks$COMP_SPECIES))

for (i in 1:length(species)) {
  
    tiles <- syn_blocks %>%
        filter(COMP_SPECIES == species[i]) %>%
        mutate(chr = paste("lg", MAP_LG, sep=''), 
               start = as.integer(MAP_START * 500000), 
               end = as.integer(MAP_STOP * 500000)) %>%
        select(chr, start, end)
    
    write.table(tiles, file = paste("data/SYNTENY/", species[i], ".synblocks", sep = ''), 
                sep = "\t", quote = FALSE, 
                row.names = FALSE, col.names = FALSE)
}

```

### Write tile track files for synteny mapped loci

Format dataframe with information on synteny mapped loci to be able to filter individuals data sets.

```{r}

# fix formating of locus names
temp <- filter(syn_loci, grepl("gi*", LOCUS)) %>%
  separate(LOCUS, c("A", "B", "C", "D"), sep = "\\|") %>%
  rename(LOCUS = B) %>%
  select(-A, -C, -D)

# add back into synteny mapped loci data frame
syn_loci <- filter(syn_loci, !grepl("gi*", LOCUS))
syn_loci <- bind_rows(syn_loci, temp)

flounderloci_IDs$LOCUS <- as.character(flounderloci_IDs$LOCUS)

# add accession numbers
syn_loci <- left_join(syn_loci, flounderloci_IDs)

nrow(syn_loci)
nrow(flounderloci_IDs)

# write file
write.table(syn_loci, "results/syn-mapped.loci", 
            col.names = TRUE, row.names = FALSE,
            sep = "\t", quote = FALSE)

```

**Williams et al 2003: Stress response in European flounder**

```{r}

# Williams et al 2003 stress resonse
stress_response <- read.delim("data/SYNTENY/Stress_response_W03.loci",
                             header = TRUE, stringsAsFactors = FALSE, sep = "\t")

# add map data locus info
stress_response <- left_join(stress_response, syn_loci)

count(stress_response, is.na(LOCUS))

# write results
write.table(stress_response, "results/stress_response.loci",
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

# Write a tile file for the synteny mapped loci
stress_response <- stress_response %>%
    filter(!is.na(LOCUS)) %>%
    mutate(chr = paste("lg", LG, sep=''), 
           start = as.integer(LEFT_POS * 500000), 
           end = as.integer(RIGHT_POS * 500000)) %>%
    select(chr, start, end)

write.table(stress_response, file = "data/SYNTENY/stress_response.syn", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)
```

**Williams et al. 2007 Estradiol exposure resonse** 

```{r}

# Williams et al. 2007 gene expression exposed to estradiol
estradiol_response <- read.delim("data/SYNTENY/Estradiol_response_W07.loci",
                                 header = TRUE, stringsAsFactors = FALSE)

# add map data locus info
estradiol_response <- left_join(estradiol_response, syn_loci)

count(estradiol_response, is.na(LOCUS))

# write results
write.table(estradiol_response, "results/estradiol_response.loci",
            col.names = TRUE, row.names = FALSE, quote = FALSE)

# Write a tile file for the synteny mapped loci
estradiol_response <- estradiol_response %>%
    filter(!is.na(LOCUS)) %>%
    mutate(chr = paste("lg", LG, sep=''), 
           start = as.integer(LEFT_POS * 500000), 
           end = as.integer(RIGHT_POS * 500000)) %>%
    select(chr, start, end)

write.table(estradiol_response, file = "data/SYNTENY/estradiol_response.syn", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

```

**Larsen et al. 2007 Salinity differences**

```{r}

# Larsen et al. 2007 gene expression salinity differences
salinity_response <- read.delim("data/SYNTENY/Salinity_response_L07.loci",
                                header = TRUE, stringsAsFactors = FALSE, sep = "\t")

# add map data locus info
salinity_response <- left_join(salinity_response, syn_loci)

count(salinity_response, is.na(LOCUS))

# write results
write.table(salinity_response, "results/salinity_response.loci",
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

# Write a tile file for the synteny mapped loci
salinity_response <- salinity_response %>%
    filter(!is.na(LOCUS)) %>%
    mutate(chr = paste("lg", LG, sep=''), 
           start = as.integer(LEFT_POS * 500000), 
           end = as.integer(RIGHT_POS * 500000)) %>%
    select(chr, start, end)

write.table(salinity_response, file = "data/SYNTENY/salinity_response.syn", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

```

**Falciani et al. 2008 Pollutants response**

```{r}

# Faciani et al. 2008 gene expression exposed to pollutants (lab and wild populations)
pollutants_response <- read.delim("data/SYNTENY/Pollutants_response_F08.loci",
                                  header = TRUE, stringsAsFactors = FALSE)

# add map data locus info
pollutants_response <- left_join(pollutants_response, syn_loci)

count(pollutants_response, is.na(LOCUS))

# write results
write.table(pollutants_response, "results/pollutants_response.loci",
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

# Write a tile file for the synteny mapped loci
pollutants_response <- pollutants_response %>%
    filter(!is.na(LOCUS)) %>%
    mutate(chr = paste("lg", LG, sep=''), 
           start = as.integer(LEFT_POS * 500000), 
           end = as.integer(RIGHT_POS * 500000)) %>%
    select(chr, start, end)

write.table(pollutants_response, file = "data/SYNTENY/pollutants_response.syn", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

```

**Song et al 2012. Linkage map japanese flounder**

```{r}

# microsatellite in linkage map (Japanese flounder)
Japfl_linkmapI <- read.delim("data/SYNTENY/LinkMapI.loci",
                            header = TRUE, stringsAsFactors = FALSE)

# add map data locus info
Japfl_linkmapI <- left_join(Japfl_linkmapI, syn_loci)

count(Japfl_linkmapI, is.na(LOCUS))

# write results
write.table(Japfl_linkmapI, "results/Japfl_linkmapI.loci",
            col.names = TRUE, row.names = FALSE, quote = FALSE)

# Write a tile file for the synteny mapped loci
Japfl_linkmapI <- Japfl_linkmapI %>%
    filter(!is.na(LOCUS)) %>%
    mutate(chr = paste("lg", LG, sep=''), 
           start = as.integer(LEFT_POS * 500000), 
           end = as.integer(RIGHT_POS * 500000)) %>%
    select(chr, start, end)

write.table(Japfl_linkmapI, file = "data/SYNTENY/Japfl_linkmapI.syn", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

```

**Castano-Sanchez et al. 2010 Linkage map japanese flounder**

```{r}

# linkage map Japanese flounder
Japfl_linkmapII <- read.delim("data/SYNTENY/LinkMapII.loci",
                               header = TRUE, stringsAsFactors = FALSE)

# add map data locus info
Japfl_linkmapII <- left_join(Japfl_linkmapII, syn_loci)

count(Japfl_linkmapII, is.na(LOCUS))

# write results
write.table(Japfl_linkmapII, "results/Japfl_linkmapII.loci",
            col.names = TRUE, row.names = FALSE, quote = FALSE)

# Write a tile file for the synteny mapped loci
Japfl_linkmapII <- Japfl_linkmapII %>%
    filter(!is.na(LOCUS)) %>%
    mutate(chr = paste("lg", LG, sep=''), 
           start = as.integer(LEFT_POS * 500000), 
           end = as.integer(RIGHT_POS * 500000)) %>%
    select(chr, start, end)

write.table(Japfl_linkmapII, file = "data/SYNTENY/Japfl_linkmapII.syn", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

```

### Write tile track files for QTL loci

```{r}

# data frame with synteny mapped loci
View(syn_loci)

# QTL 1 ----

# Fiuji et al 2006. Identification of a single major genetic locus controlling the resistance to lymphocystis disease in Japanese flounder (Paralichthys olivaceus)

# QTL for resistance to lymphocystis disease
# associated markers: Poli9-8TUF (AB037989), 

# linkage map Japanese flounder
qtl_lymph <- read.delim("data/SYNTENY/LinkMapII.loci",
                               header = TRUE, stringsAsFactors = FALSE)

# add map data locus info & filter for qtl-associated locus
qtl_lymph <- left_join(qtl_lymph, syn_loci) %>%
    filter(ACCESSION == "AB037989")

# Write a tile file
qtl_lymph <- qtl_lymph %>%
    filter(!is.na(LOCUS)) %>%
    mutate(chr = paste("lg", LG, sep=''), 
           start = as.integer(LEFT_POS * 500000), 
           end = as.integer(RIGHT_POS * 500000)) %>%
    select(chr, start, end)

write.table(qtl_lymph, file = "data/SYNTENY/qtl_lymph.syn", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

# QTL 2 ----

# Song et al 2012. Construction of High-Density Genetic Linkage Maps and Mapping of Growth-Related Quantitative Trail Loci in the Japanese Flounder (Paralichthys olivaceus)

# QTL growth
# associated markers: 
# qWe-f14 scaffold388_40956 (JN901322)
# qWi-f14-1 scaffold687_4181 (JN901305)
# qWi-f14-2 scaffold485_47120 (JN901297)
# qWi-m14 scaffold558_51720 (JN901325)

# linkage map Japanese flounder
qtl_growth <- read.delim("data/SYNTENY/LinkMapI.loci",
                               header = TRUE, stringsAsFactors = FALSE)

# add map data locus info & filter for qtl-associated locus
qtl_growth <- left_join(qtl_growth, syn_loci) %>%
    filter(ACCESSION %in% c("JN901322", "JN901305", "JN901297", "JN901325"))

# Write a tile file
qtl_growth <- qtl_growth %>%
    filter(!is.na(LOCUS)) %>%
    mutate(chr = paste("lg", LG, sep=''), 
           start = as.integer(LEFT_POS * 500000), 
           end = as.integer(RIGHT_POS * 500000)) %>%
    select(chr, start, end)

write.table(qtl_growth, file = "data/SYNTENY/qtl_growth.syn", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

# QTL 3 ----

# Wang et al 2016. Quantitative trait loci detection of Edwardsiella tarda resistance in Japanese flounder Paralichthys olivaceus using bulked segregant analysis

# QTL E. tarda resistance
# associated markers: 
# scaffold404_21589	JN900913
# scaffold404_21594	JN900903
# scaffold270_13812	JN900916

# linkage map Japanese flounder
qtl_Etarda<- read.delim("data/SYNTENY/LinkMapI.loci",
                               header = TRUE, stringsAsFactors = FALSE)

# add map data locus info & filter for qtl-associated locus
qtl_Etarda <- left_join(qtl_Etarda, syn_loci) %>%
    filter(ACCESSION %in% c("JN900913", "JN900903", "JN900916"))

# Write a tile file
qtl_Etarda <- qtl_Etarda %>%
    filter(!is.na(LOCUS)) %>%
    mutate(chr = paste("lg", LG, sep=''), 
           start = as.integer(LEFT_POS * 500000), 
           end = as.integer(RIGHT_POS * 500000)) %>%
    select(chr, start, end)

write.table(qtl_Etarda, file = "data/SYNTENY/qtl_Etarda.syn", 
            sep = "\t", quote = FALSE, 
            row.names = FALSE, col.names = FALSE)

```

## Write link files

Link file for synteny blocks dlab and SFL.

```{r}

# Format file
dlab_synblocks <- syn_blocks %>%
  filter(COMP_SPECIES == "dlab") %>%
  mutate(lg = paste("lg", MAP_LG, sep=''), 
         lgstart = as.integer(MAP_START * 500000),
         lgend = as.integer(MAP_STOP * 500000),
         chr = COMP_CHR,
         chrstart = COMP_START,
         chrend = COMP_END) %>%
  select(lg, lgstart, lgend, chr, chrstart, chrend)

write.table(dlab_synblocks, file = "data/SYNTENY/dlab.syn", 
                         sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Link file for synteny blocks gacu and SFL.

```{r}

# Format file
dlab_synblocks <- syn_blocks %>%
  filter(COMP_SPECIES == "gacu") %>%
  mutate(lg = paste("lg", MAP_LG, sep=''), 
         lgstart = as.integer(MAP_START * 500000),
         lgend = as.integer(MAP_STOP * 500000),
         chr = COMP_CHR,
         chrstart = COMP_START,
         chrend = COMP_END) %>%
  select(lg, lgstart, lgend, chr, chrstart, chrend) %>%
    group_by(lg)

write.table(dlab_synblocks, file = "data/SYNTENY/gacu.syn", 
                         sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Link file for synteny blocks lcal and SFL.

```{r}

# Format file
dlab_synblocks <- syn_blocks %>%
  filter(COMP_SPECIES == "lcal") %>%
  mutate(lg = paste("lg", MAP_LG, sep=''), 
         lgstart = as.integer(MAP_START * 500000),
         lgend = as.integer(MAP_STOP * 500000),
         chr = COMP_CHR,
         chrstart = COMP_START,
         chrend = COMP_END) %>%
  select(lg, lgstart, lgend, chr, chrstart, chrend) %>%
    group_by(lg)

write.table(dlab_synblocks, file = "data/SYNTENY/lcal.syn", 
                         sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Link file for synteny blocks onil and SFL.

```{r}

# Format file
dlab_synblocks <- syn_blocks %>%
  filter(COMP_SPECIES == "onil") %>%
  mutate(lg = paste("lg", MAP_LG, sep=''), 
         lgstart = as.integer(MAP_START * 500000),
         lgend = as.integer(MAP_STOP * 500000),
         chr = COMP_CHR,
         chrstart = COMP_START,
         chrend = COMP_END) %>%
  select(lg, lgstart, lgend, chr, chrstart, chrend) %>%
    group_by(lg)

write.table(dlab_synblocks, file = "data/SYNTENY/onil.syn", 
                         sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Link file for synteny blocks tnig and SFL.

```{r}

# Format file
dlab_synblocks <- syn_blocks %>%
  filter(COMP_SPECIES == "tnig") %>%
  mutate(lg = paste("lg", MAP_LG, sep=''), 
         lgstart = as.integer(MAP_START * 500000),
         lgend = as.integer(MAP_STOP * 500000),
         chr = COMP_CHR,
         chrstart = COMP_START,
         chrend = COMP_END) %>%
  select(lg, lgstart, lgend, chr, chrstart, chrend) %>%
    group_by(lg)

write.table(dlab_synblocks, file = "data/SYNTENY/tnig.syn", 
                         sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

Link file for synteny blocks trub and SFL.

```{r}

# Format file
dlab_synblocks <- syn_blocks %>%
  filter(COMP_SPECIES == "trub") %>%
  mutate(lg = paste("lg", MAP_LG, sep=''), 
         lgstart = as.integer(MAP_START * 500000),
         lgend = as.integer(MAP_STOP * 500000),
         chr = COMP_CHR,
         chrstart = COMP_START,
         chrend = COMP_END) %>%
  select(lg, lgstart, lgend, chr, chrstart, chrend) %>%
    group_by(lg)

write.table(dlab_synblocks, file = "data/SYNTENY/trub.syn", 
                         sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

# Run circos:

## Plot synteny blocks (inside) and SNP loci (outside):

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_blocks.conf -outputdir fig -outputfile synteny-blocks')

```

## Plot comparison location of synteny blocks 

**lcal vs SFL**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_lcal.conf -outputdir fig -outputfile synteny-comp.lcal')

```

**dlab vs SFL**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_dlab.conf -outputdir fig -outputfile synteny-comp.dlab')

```

**gacu vs SFL**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_gacu.conf -outputdir fig -outputfile synteny-comp.gacu')

```

**onil vs SFL**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_onil.conf -outputdir fig -outputfile synteny-comp.onil')

```

**tnig vs SFL**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_tnig.conf -outputdir fig -outputfile synteny-comp.tnig')

```

**trub vs SFL**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_trub.conf -outputdir fig -outputfile synteny-comp.trub')

```

## Plot synteny mapped loci

All loci: exposure studies (European flounder) and linkage maps (Japanese flounder)

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_loci.conf -outputdir fig -outputfile synt_loci')

```

Exposure studies: **Williams et al 2003: Stress response in European flounder**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_stress.conf -outputdir fig -outputfile synt_stress')

```

Exposure studies: **Williams et al. 2007 Estradiol exposure resonse**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_estradiol.conf -outputdir fig -outputfile synt_estradiol')

```

Exposure studies: **Larsen et al. 2007 Salinity differences**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_salinity.conf -outputdir fig -outputfile synt_salinity')

```

Exposure studies: **Falciani et al. 2008 Pollutants response**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_pollutants.conf -outputdir fig -outputfile synt_pollutants')

```

Exposure studies: **Pollutants and Salinity response European flounder** & **QTL Japanese flounder**

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_salinity_pollutants.conf -outputdir fig -outputfile synt_salinity_pollutants')

```

Linkage maps (Japanese flounder) w/qtl loci

```{r}

system('perl "scr/circos-0.66/bin/circos" -conf data/SYNTENY/circos-synt_loci-japfl.conf -outputdir fig -outputfile synt_loci-japfl')

```


# Analyze distribution of synteny mapped loci

## Genes identified in response to exposure to pollutants

```{r}

# pollutants_response <- read.table("results/pollutants_response.loci",
#                          header = TRUE, stringsAsFactors = FALSE)

poll_mapped <- pollutants_response %>%
    filter(!is.na(LEFT_LOCUS)) %>%
    arrange(LG, LEFT_POS) %>%
    unite(POS, LG, LEFT_POS, sep = "_",remove = FALSE)

poll_LG <- count(poll_mapped, LG)

write.table(poll_LG, "results/per_LG_pollution.mapped", 
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

poll_stacks <- count(poll_mapped, POS)

write.table(poll_stacks, "results/per_loc_pollution.mapped", 
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

temp <- filter(poll_stacks, n >= 10)

poll_loci <- poll_mapped %>%
    filter(POS %in% temp$POS)

poll_gene <- count(poll_loci, GENE_ID)

write.table(poll_gene, "results/genes_pollution.mapped", 
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

poll_GO <- count(poll_loci, GO_PROCESS)

write.table(poll_GO, "results/GO_poll.mapped", 
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

```

## Genes identified in response to salinity differences in habitat, reciprocal translocations

```{r}

# salinity_response <- read.table("results/salinity_response.loci",
#                          header = TRUE, stringsAsFactors = FALSE)

salin_mapped <- salinity_response %>%
    filter(!is.na(LEFT_LOCUS)) %>%
    arrange(LG, LEFT_POS) %>%
    unite(POS, LG, LEFT_POS, sep = "_",remove = FALSE)

salin_LG <- count(salin_mapped, LG)

write.table(salin_LG, "results/per_LG_salinution.mapped", 
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

salin_stacks <- count(salin_mapped, POS)

write.table(salin_stacks, "results/per_loc_salinution.mapped", 
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

temp <- filter(salin_stacks, n >= 10)

salin_loci <- salin_mapped %>%
    filter(POS %in% temp$POS)

salin_gene <- count(salin_loci, GENE_ID)

write.table(salin_gene, "results/genes_salinution.mapped", 
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

salin_GO <- count(salin_loci, GO_PROCESS)

write.table(salin_GO, "results/GO_salin.mapped", 
            col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")

```